{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of blasticidin samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Peak Finding\n",
    "We keep this code here for reference, analysis was done in the cluster <br>\n",
    "First we merge all the blasticidin samples together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "samtools merge -f $TMPDIR/out.bam  \\\n",
    "BSF_03/res4/BSF_03/BSF_03_sorted.bam \\\n",
    "BSF_04/res4/BSF_04/BSF_04_sorted.bam \\\n",
    "BSF_05/res4/BSF_05/BSF_05_sorted.bam \\\n",
    "\n",
    "samtools sort -@ 10 -o $TMPDIR/BSF_B_merge.bam $TMPDIR/out.bam\n",
    "rm $TMPDIR/out.bam\n",
    "samtools index $TMPDIR/BSF_B_merge.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#copy the bam file for the control samples\n",
    "!cp BSF_01/res4/BSF_01/BSF_01_sorted.bam $TMPDIR/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#call peaks using B samples bam files\n",
    "#and control bam file\n",
    "#These parameters were found after severel attempts of try/error\n",
    "!mkdir -p merge_macs_BSF_B\n",
    "!macs2 callpeak -t $TMPDIR/BSF_B_merge.bam -f BAMPE \\\n",
    "-c $TMPDIR/BSF_01_sorted.bam \\\n",
    "--min-length 300 --max-gap 1 --broad --nomodel --keep-dup all\\\n",
    "-q 1 --broad-cutoff 1 --broad --llocal 3000 --outdir 'merge_macs_BSF_B'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>22346</td>\n",
       "      <td>23528</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>26393</td>\n",
       "      <td>26954</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>27519</td>\n",
       "      <td>28717</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>44970</td>\n",
       "      <td>45408</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>50065</td>\n",
       "      <td>50477</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr  start    end  coverage\n",
       "0  11L3_v3  22346  23528         1\n",
       "1  11L3_v3  26393  26954         1\n",
       "2  11L3_v3  27519  28717         1\n",
       "3  11L3_v3  44970  45408         1\n",
       "4  11L3_v3  50065  50477         1"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#load peak results and make a bed files\n",
    "#set coverage to one for visualization\n",
    "import pandas as pd\n",
    "indf = pd.read_csv('merge_macs_BSF_B/NA_peaks.xls',sep='\\t',comment='#')\n",
    "#GeneID\tChr\tStart\tEnd\tStrand\n",
    "indf = indf[list(indf.columns[0:3])]\n",
    "indf['coverage']=1\n",
    "#indf.columns = ['GeneID','Chr','Start','End']\n",
    "indf.to_csv('merge_macs_BSF_B/merge_macs_BSF_B.bed',sep='\\t',index=False,header=False)\n",
    "indf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NA_peak_1</td>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>22346</td>\n",
       "      <td>23528</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NA_peak_2</td>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>26393</td>\n",
       "      <td>26954</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NA_peak_3</td>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>27519</td>\n",
       "      <td>28717</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NA_peak_4</td>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>44970</td>\n",
       "      <td>45408</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NA_peak_5</td>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>50065</td>\n",
       "      <td>50477</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      GeneID      Chr  Start    End Strand\n",
       "0  NA_peak_1  11L3_v3  22346  23528      .\n",
       "1  NA_peak_2  11L3_v3  26393  26954      .\n",
       "2  NA_peak_3  11L3_v3  27519  28717      .\n",
       "3  NA_peak_4  11L3_v3  44970  45408      .\n",
       "4  NA_peak_5  11L3_v3  50065  50477      ."
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#we make SAF file from the peak finding alghritm (MACS2) to use with feature counts\n",
    "indf = pd.read_csv('merge_macs_BSF_B/NA_peaks.xls',sep='\\t',comment='#')\n",
    "#GeneID\tChr\tStart\tEnd\tStrand\n",
    "indf = indf[[indf.columns[-1]] + list(indf.columns[0:3])]\n",
    "indf.columns = ['GeneID','Chr','Start','End']\n",
    "indf['Strand']='.'\n",
    "indf.to_csv('merge_macs_BSF_B/merge_macs_BSF_B.SAF',sep='\\t',index=False)\n",
    "indf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Counting\n",
    "We count, for all amples, the total reads and the reads with the barcode <br>in the 4 different orientation (F,R,FR,RR)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#we now count the read pairs for each B samples and barcoded B samples\n",
    "!featureCounts -p -B -C -M -O -T 8 -F SAF -a 'merge_macs_BSF_B/merge_macs_BSF_B.SAF' \\\n",
    "-o 'merge_macs_BSF_B/merge_macs_BSF_B_counts.txt' \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_01/res4/BSF_01/BSF_01_sorted.bam \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_03/res4/BSF_03/BSF_03_sorted.bam \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_04/res4/BSF_04/BSF_04_sorted.bam \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_05/res4/BSF_05/BSF_05_sorted.bam \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_01/res4/BSF_01/BSF_01_sorted_F.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_01/res4/BSF_01/BSF_01_sorted_R.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_01/res4/BSF_01/BSF_01_sorted_FR.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_01/res4/BSF_01/BSF_01_sorted_RR.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_03/res4/BSF_03/BSF_03_sorted_F.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_03/res4/BSF_03/BSF_03_sorted_R.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_03/res4/BSF_03/BSF_03_sorted_FR.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_03/res4/BSF_03/BSF_03_sorted_RR.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_04/res4/BSF_04/BSF_04_sorted_F.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_04/res4/BSF_04/BSF_04_sorted_R.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_04/res4/BSF_04/BSF_04_sorted_FR.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_04/res4/BSF_04/BSF_04_sorted_RR.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_05/res4/BSF_05/BSF_05_sorted_F.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_05/res4/BSF_05/BSF_05_sorted_R.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_05/res4/BSF_05/BSF_05_sorted_FR.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_05/res4/BSF_05/BSF_05_sorted_RR.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in the peak counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>BSF_01</th>\n",
       "      <th>BSF_03</th>\n",
       "      <th>BSF_04</th>\n",
       "      <th>BSF_05</th>\n",
       "      <th>BSF_01_F</th>\n",
       "      <th>...</th>\n",
       "      <th>BSF_03_FR</th>\n",
       "      <th>BSF_03_RR</th>\n",
       "      <th>BSF_04_F</th>\n",
       "      <th>BSF_04_R</th>\n",
       "      <th>BSF_04_FR</th>\n",
       "      <th>BSF_04_RR</th>\n",
       "      <th>BSF_05_F</th>\n",
       "      <th>BSF_05_R</th>\n",
       "      <th>BSF_05_FR</th>\n",
       "      <th>BSF_05_RR</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NA_peak_1</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>22346</td>\n",
       "      <td>23528</td>\n",
       "      <td>.</td>\n",
       "      <td>1183</td>\n",
       "      <td>221</td>\n",
       "      <td>689</td>\n",
       "      <td>470</td>\n",
       "      <td>354</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>23</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_2</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>26393</td>\n",
       "      <td>26954</td>\n",
       "      <td>.</td>\n",
       "      <td>562</td>\n",
       "      <td>47</td>\n",
       "      <td>94</td>\n",
       "      <td>35</td>\n",
       "      <td>13</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_3</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>27519</td>\n",
       "      <td>28717</td>\n",
       "      <td>.</td>\n",
       "      <td>1199</td>\n",
       "      <td>25</td>\n",
       "      <td>189</td>\n",
       "      <td>282</td>\n",
       "      <td>422</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_4</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>44970</td>\n",
       "      <td>45408</td>\n",
       "      <td>.</td>\n",
       "      <td>439</td>\n",
       "      <td>48</td>\n",
       "      <td>225</td>\n",
       "      <td>57</td>\n",
       "      <td>100</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_5</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>50065</td>\n",
       "      <td>50477</td>\n",
       "      <td>.</td>\n",
       "      <td>413</td>\n",
       "      <td>124</td>\n",
       "      <td>206</td>\n",
       "      <td>83</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Chr  Start    End Strand  Length  BSF_01  BSF_03  BSF_04  \\\n",
       "Geneid                                                                    \n",
       "NA_peak_1  11L3_v3  22346  23528      .    1183     221     689     470   \n",
       "NA_peak_2  11L3_v3  26393  26954      .     562      47      94      35   \n",
       "NA_peak_3  11L3_v3  27519  28717      .    1199      25     189     282   \n",
       "NA_peak_4  11L3_v3  44970  45408      .     439      48     225      57   \n",
       "NA_peak_5  11L3_v3  50065  50477      .     413     124     206      83   \n",
       "\n",
       "           BSF_05  BSF_01_F  ...  BSF_03_FR  BSF_03_RR  BSF_04_F  BSF_04_R  \\\n",
       "Geneid                       ...                                             \n",
       "NA_peak_1     354         5  ...         14         23         6         1   \n",
       "NA_peak_2      13         4  ...          0          0         2         0   \n",
       "NA_peak_3     422         0  ...         11          0         1         5   \n",
       "NA_peak_4     100         2  ...          4          1         1         1   \n",
       "NA_peak_5     100         0  ...         22          0         0        12   \n",
       "\n",
       "           BSF_04_FR  BSF_04_RR  BSF_05_F  BSF_05_R  BSF_05_FR  BSF_05_RR  \n",
       "Geneid                                                                     \n",
       "NA_peak_1          2         11         3         0          3          8  \n",
       "NA_peak_2          0          0         1         0          0          0  \n",
       "NA_peak_3         15          0         2        14         23          0  \n",
       "NA_peak_4          0          0         2         2          1          0  \n",
       "NA_peak_5          7          0         0         9          7          0  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#get the counts of all samples\n",
    "df = pd.read_csv('merge_macs_BSF_B/merge_macs_BSF_B_counts.txt',comment='#',sep='\\t',index_col=[0])\n",
    "counts_col = list(df.columns[5:])\n",
    "new_names = [n.split('/')[-1].split('.')[0].replace('_sorted','') for n in counts_col]\n",
    "new_names = dict(zip(counts_col,new_names))\n",
    "df = df.rename(new_names,axis=1)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NA_peak_1</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>22346</td>\n",
       "      <td>23528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_2</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>26393</td>\n",
       "      <td>26954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_3</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>27519</td>\n",
       "      <td>28717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_4</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>44970</td>\n",
       "      <td>45408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_5</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>50065</td>\n",
       "      <td>50477</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Chr  Start    End\n",
       "Geneid                          \n",
       "NA_peak_1  11L3_v3  22346  23528\n",
       "NA_peak_2  11L3_v3  26393  26954\n",
       "NA_peak_3  11L3_v3  27519  28717\n",
       "NA_peak_4  11L3_v3  44970  45408\n",
       "NA_peak_5  11L3_v3  50065  50477"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Prepare a final table\n",
    "final_table=pd.DataFrame(index=df.index)\n",
    "final_table['Chr']=df.Chr\n",
    "final_table['Start']=df.Start\n",
    "final_table['End']=df.End\n",
    "final_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#[n for n in df.columns if '_03' in n]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sum up the barcoded reads\n",
    "we create one column with the sum of the barcoded reads in <br>\n",
    "forward (F/RR) and reverse complement (R/RR) orientation <br>\n",
    "using control and treatments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Control_total</th>\n",
       "      <th>Control_F</th>\n",
       "      <th>Control_R</th>\n",
       "      <th>Treatment_total</th>\n",
       "      <th>Treatment_F</th>\n",
       "      <th>Treatment_R</th>\n",
       "      <th>Treatment_fraction_F</th>\n",
       "      <th>Treatment_fraction_R</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NA_peak_1</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>22346</td>\n",
       "      <td>23528</td>\n",
       "      <td>221</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>1513</td>\n",
       "      <td>54</td>\n",
       "      <td>25</td>\n",
       "      <td>0.683544</td>\n",
       "      <td>0.316456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_2</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>26393</td>\n",
       "      <td>26954</td>\n",
       "      <td>47</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>142</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_3</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>27519</td>\n",
       "      <td>28717</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>893</td>\n",
       "      <td>5</td>\n",
       "      <td>71</td>\n",
       "      <td>0.065789</td>\n",
       "      <td>0.934211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_4</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>44970</td>\n",
       "      <td>45408</td>\n",
       "      <td>48</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>382</td>\n",
       "      <td>6</td>\n",
       "      <td>13</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.684211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_5</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>50065</td>\n",
       "      <td>50477</td>\n",
       "      <td>124</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>389</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Chr  Start    End  Control_total  Control_F  Control_R  \\\n",
       "Geneid                                                                  \n",
       "NA_peak_1  11L3_v3  22346  23528            221          8          9   \n",
       "NA_peak_2  11L3_v3  26393  26954             47          4          0   \n",
       "NA_peak_3  11L3_v3  27519  28717             25          0          1   \n",
       "NA_peak_4  11L3_v3  44970  45408             48          2          0   \n",
       "NA_peak_5  11L3_v3  50065  50477            124          4         12   \n",
       "\n",
       "           Treatment_total  Treatment_F  Treatment_R  Treatment_fraction_F  \\\n",
       "Geneid                                                                       \n",
       "NA_peak_1             1513           54           25              0.683544   \n",
       "NA_peak_2              142           14            0              1.000000   \n",
       "NA_peak_3              893            5           71              0.065789   \n",
       "NA_peak_4              382            6           13              0.315789   \n",
       "NA_peak_5              389            0           72              0.000000   \n",
       "\n",
       "           Treatment_fraction_R  \n",
       "Geneid                           \n",
       "NA_peak_1              0.316456  \n",
       "NA_peak_2              0.000000  \n",
       "NA_peak_3              0.934211  \n",
       "NA_peak_4              0.684211  \n",
       "NA_peak_5              1.000000  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#prepare the control\n",
    "final_table['Control_total']=df['BSF_01']\n",
    "final_table['Control_F']=df['BSF_01_F']+df['BSF_01_RR']\n",
    "final_table['Control_R']=df['BSF_01_R']+df['BSF_01_FR']\n",
    "#prepare treatment\n",
    "final_table['Treatment_total']=df['BSF_03']+df['BSF_04']+df['BSF_05']\n",
    "\n",
    "final_table['Treatment_F']=(df['BSF_03_F']+df['BSF_04_F']+df['BSF_05_F']+\n",
    "                            df['BSF_03_RR']+df['BSF_04_RR']+df['BSF_05_RR'])\n",
    "\n",
    "final_table['Treatment_R']=(df['BSF_03_R']+df['BSF_04_R']+df['BSF_05_R']+\n",
    "                            df['BSF_03_FR']+df['BSF_04_FR']+df['BSF_05_FR'])\n",
    "\n",
    "#fraction of reads in forward oritentation\n",
    "#now we try to extract the orientation of the peak\n",
    "#it will be a number between 0 and 1\n",
    "#so, for a peak, having Treatment_fraction_F ==1 \n",
    "#means all the barcoded reads where found in forward orientation\n",
    "#this is clearly an apporximantion, as sometimes is difficoult to \n",
    "#detect peack that span the full barcoded region\n",
    "final_table['Treatment_fraction_F']=final_table['Treatment_F']/(final_table['Treatment_F']+final_table['Treatment_R'])\n",
    "#by definition\n",
    "final_table['Treatment_fraction_R']=1-final_table['Treatment_fraction_F']\n",
    "final_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Control_total</th>\n",
       "      <th>Control_F</th>\n",
       "      <th>Control_R</th>\n",
       "      <th>Treatment_total</th>\n",
       "      <th>Treatment_F</th>\n",
       "      <th>Treatment_R</th>\n",
       "      <th>Treatment_fraction_F</th>\n",
       "      <th>Treatment_fraction_R</th>\n",
       "      <th>Treatment_total_corrected_F</th>\n",
       "      <th>Treatment_total_corrected_R</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NA_peak_1</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>22346</td>\n",
       "      <td>23528</td>\n",
       "      <td>221</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>1513</td>\n",
       "      <td>54</td>\n",
       "      <td>25</td>\n",
       "      <td>0.683544</td>\n",
       "      <td>0.316456</td>\n",
       "      <td>1034.202532</td>\n",
       "      <td>478.797468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_2</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>26393</td>\n",
       "      <td>26954</td>\n",
       "      <td>47</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>142</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>142.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_3</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>27519</td>\n",
       "      <td>28717</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>893</td>\n",
       "      <td>5</td>\n",
       "      <td>71</td>\n",
       "      <td>0.065789</td>\n",
       "      <td>0.934211</td>\n",
       "      <td>58.750000</td>\n",
       "      <td>834.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_4</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>44970</td>\n",
       "      <td>45408</td>\n",
       "      <td>48</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>382</td>\n",
       "      <td>6</td>\n",
       "      <td>13</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.684211</td>\n",
       "      <td>120.631579</td>\n",
       "      <td>261.368421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_5</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>50065</td>\n",
       "      <td>50477</td>\n",
       "      <td>124</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>389</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>389.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Chr  Start    End  Control_total  Control_F  Control_R  \\\n",
       "Geneid                                                                  \n",
       "NA_peak_1  11L3_v3  22346  23528            221          8          9   \n",
       "NA_peak_2  11L3_v3  26393  26954             47          4          0   \n",
       "NA_peak_3  11L3_v3  27519  28717             25          0          1   \n",
       "NA_peak_4  11L3_v3  44970  45408             48          2          0   \n",
       "NA_peak_5  11L3_v3  50065  50477            124          4         12   \n",
       "\n",
       "           Treatment_total  Treatment_F  Treatment_R  Treatment_fraction_F  \\\n",
       "Geneid                                                                       \n",
       "NA_peak_1             1513           54           25              0.683544   \n",
       "NA_peak_2              142           14            0              1.000000   \n",
       "NA_peak_3              893            5           71              0.065789   \n",
       "NA_peak_4              382            6           13              0.315789   \n",
       "NA_peak_5              389            0           72              0.000000   \n",
       "\n",
       "           Treatment_fraction_R  Treatment_total_corrected_F  \\\n",
       "Geneid                                                         \n",
       "NA_peak_1              0.316456                  1034.202532   \n",
       "NA_peak_2              0.000000                   142.000000   \n",
       "NA_peak_3              0.934211                    58.750000   \n",
       "NA_peak_4              0.684211                   120.631579   \n",
       "NA_peak_5              1.000000                     0.000000   \n",
       "\n",
       "           Treatment_total_corrected_R  \n",
       "Geneid                                  \n",
       "NA_peak_1                   478.797468  \n",
       "NA_peak_2                     0.000000  \n",
       "NA_peak_3                   834.250000  \n",
       "NA_peak_4                   261.368421  \n",
       "NA_peak_5                   389.000000  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#now we adjust the total reads, that we assume having an orientation\n",
    "#with the same proportion of the barcoded reads\n",
    "final_table['Treatment_total_corrected_F']=final_table['Treatment_total']*final_table['Treatment_fraction_F']\n",
    "final_table['Treatment_total_corrected_R']=final_table['Treatment_total']*final_table['Treatment_fraction_R']\n",
    "final_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11L3_v3\t22346\t23528\tNA_peak_1\n",
      "11L3_v3\t26393\t26954\tNA_peak_2\n",
      "11L3_v3\t27519\t28717\tNA_peak_3\n",
      "11L3_v3\t44970\t45408\tNA_peak_4\n",
      "11L3_v3\t50065\t50477\tNA_peak_5\n",
      "11L3_v3\t54025\t54923\tNA_peak_6\n",
      "5K5_v5.1\t1983\t3503\tNA_peak_7\n",
      "5K5_v5.1\t3902\t5898\tNA_peak_8\n",
      "5K5_v5.1\t6985\t8467\tNA_peak_9\n",
      "5K5_v5.1\t10746\t11356\tNA_peak_10\n"
     ]
    }
   ],
   "source": [
    "#create a bed file from the peak start/end \n",
    "bed_file = df[['Chr', 'Start', 'End']]\n",
    "bed_file=bed_file.reset_index()\n",
    "bed_file = bed_file[['Chr', 'Start', 'End', 'Geneid']]\n",
    "bed_file=bed_file.rename({'Geneid':'Peak_Id'},axis=1)\n",
    "bed_file[['Chr', 'Start', 'End', 'Peak_Id']].to_csv('bla_peaks_final.bed', sep='\\t',\n",
    "                                                   header=False, index=False)\n",
    "!head 'bla_peaks_final.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(284, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>Tb927.9.15950</td>\n",
       "      <td>UTR_contained</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>Tb927.9.6570</td>\n",
       "      <td>UTR_contained</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>Tb927.9.6830</td>\n",
       "      <td>UTR_contained</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>Tb927.9.7500</td>\n",
       "      <td>UTR_contained</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>Tb927.9.7850</td>\n",
       "      <td>UTR_contained</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene         reason\n",
       "279  Tb927.9.15950  UTR_contained\n",
       "280   Tb927.9.6570  UTR_contained\n",
       "281   Tb927.9.6830  UTR_contained\n",
       "282   Tb927.9.7500  UTR_contained\n",
       "283   Tb927.9.7850  UTR_contained"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#From the re-annotation of the UTRs \n",
    "#we defined a set of gene that we do not belive to be real.\n",
    "#so, we want exclude those from the analysis\n",
    "bl =pd.read_csv('blacklisted gene.csv',engine='python',header=None)\n",
    "bl.columns = ['gene','reason']\n",
    "bl['gene']=[n.strip() for n in bl['gene']]\n",
    "print(bl.shape)\n",
    "bl.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>utr5</th>\n",
       "      <th>utr3</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Tb927.9.13440</td>\n",
       "      <td>Tb927_09_v5.1:2132976..2133162:r</td>\n",
       "      <td>Tb927_09_v5.1:2127479..2128238:r</td>\n",
       "      <td>Tb927_09_v5.1</td>\n",
       "      <td>2127479</td>\n",
       "      <td>2128238</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tb927.9.13430</td>\n",
       "      <td>Tb927_09_v5.1:2126776..2127309:r</td>\n",
       "      <td>Tb927_09_v5.1:2121838..2125437:r</td>\n",
       "      <td>Tb927_09_v5.1</td>\n",
       "      <td>2121838</td>\n",
       "      <td>2125437</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>3599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tb927.9.13380</td>\n",
       "      <td>Tb927_09_v5.1:2120866..2121171:r</td>\n",
       "      <td>Tb927_09_v5.1:2118125..2119605:r</td>\n",
       "      <td>Tb927_09_v5.1</td>\n",
       "      <td>2118125</td>\n",
       "      <td>2119605</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Tb927.9.13360</td>\n",
       "      <td>Tb927_09_v5.1:2117171..2117516:r</td>\n",
       "      <td>Tb927_09_v5.1:2108712..2111431:r</td>\n",
       "      <td>Tb927_09_v5.1</td>\n",
       "      <td>2108712</td>\n",
       "      <td>2111431</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Tb927.9.8905</td>\n",
       "      <td>Tb927_09_v5.1:1431796..1431807:r</td>\n",
       "      <td>Tb927_09_v5.1:1431368..1431516:r</td>\n",
       "      <td>Tb927_09_v5.1</td>\n",
       "      <td>1431368</td>\n",
       "      <td>1431516</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene_id                              utr5  \\\n",
       "0  Tb927.9.13440  Tb927_09_v5.1:2132976..2133162:r   \n",
       "1  Tb927.9.13430  Tb927_09_v5.1:2126776..2127309:r   \n",
       "2  Tb927.9.13380  Tb927_09_v5.1:2120866..2121171:r   \n",
       "3  Tb927.9.13360  Tb927_09_v5.1:2117171..2117516:r   \n",
       "4   Tb927.9.8905  Tb927_09_v5.1:1431796..1431807:r   \n",
       "\n",
       "                               utr3            Chr    Start      End Score  \\\n",
       "0  Tb927_09_v5.1:2127479..2128238:r  Tb927_09_v5.1  2127479  2128238     .   \n",
       "1  Tb927_09_v5.1:2121838..2125437:r  Tb927_09_v5.1  2121838  2125437     .   \n",
       "2  Tb927_09_v5.1:2118125..2119605:r  Tb927_09_v5.1  2118125  2119605     .   \n",
       "3  Tb927_09_v5.1:2108712..2111431:r  Tb927_09_v5.1  2108712  2111431     .   \n",
       "4  Tb927_09_v5.1:1431368..1431516:r  Tb927_09_v5.1  1431368  1431516     .   \n",
       "\n",
       "  Strand  diff  \n",
       "0      -   759  \n",
       "1      -  3599  \n",
       "2      -  1480  \n",
       "3      -  2719  \n",
       "4      -   148  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# From the re-annotation of the UTRs\n",
    "# we parse a file with UTRs coordinates\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "df = pd.read_csv('UTRs.txt',sep='\\t',header=None)\n",
    "df.columns = ['Gene_id','utr5','utr3']\n",
    "df['Chr']=[n.split(':')[0] if str(n) != 'nan' else np.nan for n in df['utr3'] ]\n",
    "df['Start']=[n.split(':')[1].split('..')[0] if str(n) != 'nan' else np.nan for n in df['utr3']]\n",
    "df['End']=[n.split(':')[1].split('..')[1] if str(n) != 'nan' else np.nan for n in df['utr3']]\n",
    "df['Score']='.'\n",
    "df['Strand']=['-' if str(n)[-1] == 'r' else '+' for n in df['utr3']]\n",
    "df=df.dropna(subset=['End'])\n",
    "df['diff'] = df['End'].astype(int)-df['Start'].astype(int)\n",
    "df=df[df['diff']>=1]\n",
    "df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#remove blacklisted genes\n",
    "df = df[~df['Gene_id'].isin(bl['gene'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#we create a bed file for the 3' UTR coords\n",
    "df[['Chr', 'Start', 'End','Gene_id', 'Score', 'Strand']].to_csv(\n",
    "    'UTR3_final.bed', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find peak / UTRs overlaps\n",
    "we use  bedtools intersect to find the ovrlapping reagions <br>\n",
    "between UTRs and Peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "bedtools intersect -a <(bedtools sort -i bla_peaks_final.bed) \\\n",
    "-b <(bedtools sort -i UTR3_final.bed) -wao > bla_peak_belongs_to_3UTR.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#use awk to find the peaks that are fully fully contained in UTRs\n",
    "!awk '$(NF-2) != \"-1\"' bla_peak_belongs_to_3UTR.bed > bla_peak_belongs_to_3UTR_filtered.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Peak_Start</th>\n",
       "      <th>Peak_End</th>\n",
       "      <th>Peak_Name</th>\n",
       "      <th>UTR_Start</th>\n",
       "      <th>UTR_End</th>\n",
       "      <th>UTR_Gene_ID</th>\n",
       "      <th>UTR_Strand</th>\n",
       "      <th>UTR_Peak_overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>803</th>\n",
       "      <td>Tb927_07_v5.1</td>\n",
       "      <td>297730</td>\n",
       "      <td>300530</td>\n",
       "      <td>NA_peak_1371</td>\n",
       "      <td>297863</td>\n",
       "      <td>297962</td>\n",
       "      <td>Tb927.7.1150</td>\n",
       "      <td>+</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome  Peak_Start  Peak_End     Peak_Name  UTR_Start  UTR_End  \\\n",
       "803  Tb927_07_v5.1      297730    300530  NA_peak_1371     297863   297962   \n",
       "\n",
       "      UTR_Gene_ID UTR_Strand  UTR_Peak_overlap  \n",
       "803  Tb927.7.1150          +                99  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#now we parse the bedtools output\n",
    "df = pd.read_csv('bla_peak_belongs_to_3UTR_filtered.bed',sep='\\t',header=None)\n",
    "df.columns = ['Chromosome','Peak_Start','Peak_End','Peak_Name',\n",
    "              'UTR_Chromosome','UTR_Start','UTR_End','UTR_Gene_ID','UTR_Score','UTR_Strand','UTR_Peak_overlap']\n",
    "del df['UTR_Score']\n",
    "del df['UTR_Chromosome']\n",
    "\n",
    "#we separate the peaks that overlap with UTRs in the forward (+) or reverse complement (-) orientation\n",
    "df_plus = df[df['UTR_Strand']=='+']\n",
    "df_minus = df[df['UTR_Strand']=='-']\n",
    "#now it is easy to discard peaks that overalp with multiple UTRs\n",
    "#we keep the most 5' if the peak is in the forward (+) orientation\n",
    "df_plus = df_plus.drop_duplicates(subset=['Peak_Name'],keep='first')\n",
    "#we keep the most 3' if the peak is in the reverse complement (+) orientation\n",
    "df_minus = df_minus.drop_duplicates(subset=['Peak_Name'],keep='last')\n",
    "df = pd.concat([df_plus,df_minus])\n",
    "df=df.sort_values(by=['Chromosome','Peak_Start','Peak_End'])\n",
    "df[df['Peak_Name']=='NA_peak_1371']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2068, 9)\n",
      "(2068, 19)\n"
     ]
    }
   ],
   "source": [
    "# now we merge the table with the peak counting\n",
    "# with the table that select the peaks overalpping with UTRs\n",
    "# we keep only the peak that overlap with a UTR\n",
    "print(df.shape)\n",
    "df = df.merge(final_table.iloc[:,3:],left_on='Peak_Name',right_index=True,how='left')\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Peak_Start</th>\n",
       "      <th>Peak_End</th>\n",
       "      <th>Peak_Name</th>\n",
       "      <th>UTR_Start</th>\n",
       "      <th>UTR_End</th>\n",
       "      <th>UTR_Gene_ID</th>\n",
       "      <th>UTR_Strand</th>\n",
       "      <th>UTR_Peak_overlap</th>\n",
       "      <th>Control_total</th>\n",
       "      <th>Control_F</th>\n",
       "      <th>Control_R</th>\n",
       "      <th>Treatment_total</th>\n",
       "      <th>Treatment_F</th>\n",
       "      <th>Treatment_R</th>\n",
       "      <th>Treatment_fraction_F</th>\n",
       "      <th>Treatment_fraction_R</th>\n",
       "      <th>Treatment_total_corrected_F</th>\n",
       "      <th>Treatment_total_corrected_R</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1857</th>\n",
       "      <td>Tb927_11_v5.1</td>\n",
       "      <td>314906</td>\n",
       "      <td>315992</td>\n",
       "      <td>NA_peak_3980</td>\n",
       "      <td>314972</td>\n",
       "      <td>316981</td>\n",
       "      <td>Tb927.11.1110</td>\n",
       "      <td>+</td>\n",
       "      <td>1020</td>\n",
       "      <td>430</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>10462</td>\n",
       "      <td>286</td>\n",
       "      <td>95</td>\n",
       "      <td>0.750656</td>\n",
       "      <td>0.249344</td>\n",
       "      <td>7853.364829</td>\n",
       "      <td>2608.635171</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chromosome  Peak_Start  Peak_End     Peak_Name  UTR_Start  UTR_End  \\\n",
       "1857  Tb927_11_v5.1      314906    315992  NA_peak_3980     314972   316981   \n",
       "\n",
       "        UTR_Gene_ID UTR_Strand  UTR_Peak_overlap  Control_total  Control_F  \\\n",
       "1857  Tb927.11.1110          +              1020            430          9   \n",
       "\n",
       "      Control_R  Treatment_total  Treatment_F  Treatment_R  \\\n",
       "1857         12            10462          286           95   \n",
       "\n",
       "      Treatment_fraction_F  Treatment_fraction_R  Treatment_total_corrected_F  \\\n",
       "1857              0.750656              0.249344                  7853.364829   \n",
       "\n",
       "      Treatment_total_corrected_R  \n",
       "1857                  2608.635171  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['UTR_Gene_ID']=='Tb927.11.1110']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Peak_Start</th>\n",
       "      <th>Peak_End</th>\n",
       "      <th>Peak_Name</th>\n",
       "      <th>UTR_Start</th>\n",
       "      <th>UTR_End</th>\n",
       "      <th>UTR_Gene_ID</th>\n",
       "      <th>UTR_Strand</th>\n",
       "      <th>UTR_Peak_overlap</th>\n",
       "      <th>Control_total</th>\n",
       "      <th>Control_F</th>\n",
       "      <th>Control_R</th>\n",
       "      <th>Treatment_total</th>\n",
       "      <th>Treatment_F</th>\n",
       "      <th>Treatment_R</th>\n",
       "      <th>Treatment_fraction_F</th>\n",
       "      <th>Treatment_fraction_R</th>\n",
       "      <th>Treatment_total_corrected_F</th>\n",
       "      <th>Treatment_total_corrected_R</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1837</th>\n",
       "      <td>Tb927_11_v5.1</td>\n",
       "      <td>125016</td>\n",
       "      <td>127658</td>\n",
       "      <td>NA_peak_3960</td>\n",
       "      <td>119846</td>\n",
       "      <td>127190</td>\n",
       "      <td>Tb927.11.510</td>\n",
       "      <td>-</td>\n",
       "      <td>2174</td>\n",
       "      <td>693</td>\n",
       "      <td>7</td>\n",
       "      <td>18</td>\n",
       "      <td>31255</td>\n",
       "      <td>105</td>\n",
       "      <td>968</td>\n",
       "      <td>0.097856</td>\n",
       "      <td>0.902144</td>\n",
       "      <td>3058.504194</td>\n",
       "      <td>28196.495806</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chromosome  Peak_Start  Peak_End     Peak_Name  UTR_Start  UTR_End  \\\n",
       "1837  Tb927_11_v5.1      125016    127658  NA_peak_3960     119846   127190   \n",
       "\n",
       "       UTR_Gene_ID UTR_Strand  UTR_Peak_overlap  Control_total  Control_F  \\\n",
       "1837  Tb927.11.510          -              2174            693          7   \n",
       "\n",
       "      Control_R  Treatment_total  Treatment_F  Treatment_R  \\\n",
       "1837         18            31255          105          968   \n",
       "\n",
       "      Treatment_fraction_F  Treatment_fraction_R  Treatment_total_corrected_F  \\\n",
       "1837              0.097856              0.902144                  3058.504194   \n",
       "\n",
       "      Treatment_total_corrected_R  \n",
       "1837                 28196.495806  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['UTR_Gene_ID']=='Tb927.11.510']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#69815-69903"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#For convinence, we create a function for \n",
    "#creating a string that can be used at TryTripDB\n",
    "#to download nucleotide sequences\n",
    "def format_download(X):\n",
    "    if X['UTR_Strand']=='-':\n",
    "        orient = 'r'\n",
    "    else:\n",
    "        orient = 'f'\n",
    "    chro =  X['Chromosome']  \n",
    "    p_start =  int(X['Peak_Start'])\n",
    "    p_end =  int(X['Peak_End'])\n",
    "    u_start =  int(X['UTR_Start'])\n",
    "    u_end =  int(X['UTR_End'])  \n",
    "    \n",
    "    start = max(p_start,u_start)\n",
    "    end = min(p_end,u_end)\n",
    "    download_string = f'{chro}:{start}..{end}:{orient}'\n",
    "    return download_string           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['download_seq']=df.apply(format_download,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#now we report if the peak\n",
    "#is fully contained in the UTR\n",
    "def is_fully_contained(X):\n",
    "    p_start =  int(X['Peak_Start'])\n",
    "    p_end =  int(X['Peak_End'])\n",
    "    u_start =  int(X['UTR_Start'])\n",
    "    u_end =  int(X['UTR_End'])\n",
    "    if (p_start>=u_start) and (p_end<=u_end):\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "df['is_fully_contained'] = df.apply(is_fully_contained,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(217, 21)"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['is_fully_contained']].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2068, 21)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Peak_Start</th>\n",
       "      <th>Peak_End</th>\n",
       "      <th>Peak_Name</th>\n",
       "      <th>UTR_Start</th>\n",
       "      <th>UTR_End</th>\n",
       "      <th>UTR_Gene_ID</th>\n",
       "      <th>UTR_Strand</th>\n",
       "      <th>UTR_Peak_overlap</th>\n",
       "      <th>Control_total</th>\n",
       "      <th>...</th>\n",
       "      <th>Control_R</th>\n",
       "      <th>Treatment_total</th>\n",
       "      <th>Treatment_F</th>\n",
       "      <th>Treatment_R</th>\n",
       "      <th>Treatment_fraction_F</th>\n",
       "      <th>Treatment_fraction_R</th>\n",
       "      <th>Treatment_total_corrected_F</th>\n",
       "      <th>Treatment_total_corrected_R</th>\n",
       "      <th>download_seq</th>\n",
       "      <th>is_fully_contained</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>69815</td>\n",
       "      <td>70745</td>\n",
       "      <td>NA_peak_118</td>\n",
       "      <td>69661</td>\n",
       "      <td>69903</td>\n",
       "      <td>Tb927.1.180</td>\n",
       "      <td>-</td>\n",
       "      <td>88</td>\n",
       "      <td>6972</td>\n",
       "      <td>...</td>\n",
       "      <td>71</td>\n",
       "      <td>15571</td>\n",
       "      <td>408</td>\n",
       "      <td>57</td>\n",
       "      <td>0.877419</td>\n",
       "      <td>0.122581</td>\n",
       "      <td>13662.296774</td>\n",
       "      <td>1908.703226</td>\n",
       "      <td>Tb927_01_v5.1:69815..69903:r</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>85792</td>\n",
       "      <td>86127</td>\n",
       "      <td>NA_peak_120</td>\n",
       "      <td>85611</td>\n",
       "      <td>85853</td>\n",
       "      <td>Tb927.1.220</td>\n",
       "      <td>-</td>\n",
       "      <td>61</td>\n",
       "      <td>3363</td>\n",
       "      <td>...</td>\n",
       "      <td>140</td>\n",
       "      <td>7286</td>\n",
       "      <td>523</td>\n",
       "      <td>154</td>\n",
       "      <td>0.772526</td>\n",
       "      <td>0.227474</td>\n",
       "      <td>5628.623338</td>\n",
       "      <td>1657.376662</td>\n",
       "      <td>Tb927_01_v5.1:85792..85853:r</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>202076</td>\n",
       "      <td>204347</td>\n",
       "      <td>NA_peak_125</td>\n",
       "      <td>204160</td>\n",
       "      <td>204417</td>\n",
       "      <td>Tb927.1.540</td>\n",
       "      <td>-</td>\n",
       "      <td>187</td>\n",
       "      <td>5250</td>\n",
       "      <td>...</td>\n",
       "      <td>173</td>\n",
       "      <td>22674</td>\n",
       "      <td>542</td>\n",
       "      <td>438</td>\n",
       "      <td>0.553061</td>\n",
       "      <td>0.446939</td>\n",
       "      <td>12540.110204</td>\n",
       "      <td>10133.889796</td>\n",
       "      <td>Tb927_01_v5.1:204160..204347:r</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>230915</td>\n",
       "      <td>232950</td>\n",
       "      <td>NA_peak_126</td>\n",
       "      <td>231675</td>\n",
       "      <td>232503</td>\n",
       "      <td>Tb927.1.700</td>\n",
       "      <td>-</td>\n",
       "      <td>828</td>\n",
       "      <td>122</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>3232</td>\n",
       "      <td>10</td>\n",
       "      <td>202</td>\n",
       "      <td>0.047170</td>\n",
       "      <td>0.952830</td>\n",
       "      <td>152.452830</td>\n",
       "      <td>3079.547170</td>\n",
       "      <td>Tb927_01_v5.1:231675..232503:r</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>263874</td>\n",
       "      <td>264334</td>\n",
       "      <td>NA_peak_129</td>\n",
       "      <td>263913</td>\n",
       "      <td>264383</td>\n",
       "      <td>Tb927.1.880</td>\n",
       "      <td>-</td>\n",
       "      <td>421</td>\n",
       "      <td>285</td>\n",
       "      <td>...</td>\n",
       "      <td>28</td>\n",
       "      <td>6872</td>\n",
       "      <td>915</td>\n",
       "      <td>7</td>\n",
       "      <td>0.992408</td>\n",
       "      <td>0.007592</td>\n",
       "      <td>6819.826464</td>\n",
       "      <td>52.173536</td>\n",
       "      <td>Tb927_01_v5.1:263913..264334:r</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome  Peak_Start  Peak_End    Peak_Name  UTR_Start  UTR_End  \\\n",
       "0  Tb927_01_v5.1       69815     70745  NA_peak_118      69661    69903   \n",
       "1  Tb927_01_v5.1       85792     86127  NA_peak_120      85611    85853   \n",
       "3  Tb927_01_v5.1      202076    204347  NA_peak_125     204160   204417   \n",
       "4  Tb927_01_v5.1      230915    232950  NA_peak_126     231675   232503   \n",
       "5  Tb927_01_v5.1      263874    264334  NA_peak_129     263913   264383   \n",
       "\n",
       "   UTR_Gene_ID UTR_Strand  UTR_Peak_overlap  Control_total  ...  Control_R  \\\n",
       "0  Tb927.1.180          -                88           6972  ...         71   \n",
       "1  Tb927.1.220          -                61           3363  ...        140   \n",
       "3  Tb927.1.540          -               187           5250  ...        173   \n",
       "4  Tb927.1.700          -               828            122  ...          4   \n",
       "5  Tb927.1.880          -               421            285  ...         28   \n",
       "\n",
       "   Treatment_total  Treatment_F  Treatment_R  Treatment_fraction_F  \\\n",
       "0            15571          408           57              0.877419   \n",
       "1             7286          523          154              0.772526   \n",
       "3            22674          542          438              0.553061   \n",
       "4             3232           10          202              0.047170   \n",
       "5             6872          915            7              0.992408   \n",
       "\n",
       "   Treatment_fraction_R  Treatment_total_corrected_F  \\\n",
       "0              0.122581                 13662.296774   \n",
       "1              0.227474                  5628.623338   \n",
       "3              0.446939                 12540.110204   \n",
       "4              0.952830                   152.452830   \n",
       "5              0.007592                  6819.826464   \n",
       "\n",
       "   Treatment_total_corrected_R                    download_seq  \\\n",
       "0                  1908.703226    Tb927_01_v5.1:69815..69903:r   \n",
       "1                  1657.376662    Tb927_01_v5.1:85792..85853:r   \n",
       "3                 10133.889796  Tb927_01_v5.1:204160..204347:r   \n",
       "4                  3079.547170  Tb927_01_v5.1:231675..232503:r   \n",
       "5                    52.173536  Tb927_01_v5.1:263913..264334:r   \n",
       "\n",
       "  is_fully_contained  \n",
       "0              False  \n",
       "1              False  \n",
       "3              False  \n",
       "4              False  \n",
       "5              False  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Chromosome', 'Peak_Start', 'Peak_End', 'Peak_Name', 'UTR_Start',\n",
       "       'UTR_End', 'UTR_Gene_ID', 'UTR_Strand', 'UTR_Peak_overlap',\n",
       "       'Control_total', 'Control_F', 'Control_R', 'Treatment_total',\n",
       "       'Treatment_F', 'Treatment_R', 'Treatment_fraction_F',\n",
       "       'Treatment_fraction_R', 'Treatment_total_corrected_F',\n",
       "       'Treatment_total_corrected_R', 'download_seq', 'is_fully_contained'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#we save now save the table for the paper\n",
    "df.to_csv('Paper_Table_Blasticidine.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sys==Python BuiltIn\n",
      "numpy==1.21.6\n",
      "pandas==1.4.2\n"
     ]
    }
   ],
   "source": [
    "# show versions of packages\n",
    "# adopted from https://stackoverflow.com/questions/40428931/package-for-listing-version-of-packages-used-in-a-jupyter-notebook\n",
    "import pkg_resources\n",
    "import types\n",
    "import sys\n",
    "def get_imports():\n",
    "    for name, val in globals().items():\n",
    "        if isinstance(val, types.ModuleType):\n",
    "            # Split ensures you get root package, \n",
    "            # not just imported function\n",
    "            name = val.__name__.split(\".\")[0]\n",
    "        elif isinstance(val, type):\n",
    "            name = val.__module__.split(\".\")[0]\n",
    "        # Some packages are weird and have different\n",
    "        # imported names vs. system/pip names. Unfortunately,\n",
    "        # there is no systematic way to get pip names from\n",
    "        # a package's imported name. You'll have to add\n",
    "        # exceptions to this list manually!\n",
    "        poorly_named_packages = {\n",
    "            \"sklearn\": \"scikit-learn\"\n",
    "        }\n",
    "        if name in poorly_named_packages.keys():\n",
    "            name = poorly_named_packages[name]\n",
    "        yield name.lower()\n",
    "imports = list(set(get_imports()))\n",
    "\n",
    "# The only way I found to get the version of the root package\n",
    "# from only the name of the package is to cross-check the names \n",
    "# of installed packages vs. imported packages\n",
    "modules = []\n",
    "for m in sys.builtin_module_names:\n",
    "    if m.lower() in imports and m !='builtins':\n",
    "        modules.append((m,'Python BuiltIn'))\n",
    "        imports.remove(m.lower())\n",
    "\n",
    "for m in pkg_resources.working_set:\n",
    "    if m.project_name.lower() in imports and m.project_name!=\"pip\":\n",
    "        modules.append((m.project_name, m.version))\n",
    "        imports.remove(m.project_name.lower())\n",
    "\n",
    "for m in sys.modules:\n",
    "    if m.lower() in imports and m !='builtins':\n",
    "        modules.append((m,'unknown'))\n",
    "\n",
    "# print('System=='+platform.system()+' '+platform.release()+'; Version=='+platform.version())\n",
    "for r in modules:\n",
    "    if 'unknown' in r[1]:\n",
    "        continue\n",
    "    print(\"{}=={}\".format(*r))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#bedtools-2.31"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python 3.10.10\n"
     ]
    }
   ],
   "source": [
    "!python --version"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
