{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "samtools merge -f $TMPDIR/out.bam  \\\n",
    "BSF_06/res4/BSF_06/BSF_06_sorted.bam \\\n",
    "BSF_07/res4/BSF_07/BSF_07_sorted.bam \\\n",
    "BSF_08/res4/BSF_08/BSF_08_sorted.bam \\\n",
    "\n",
    "samtools sort -@ 10 -o $TMPDIR/BSF_G_merge.bam $TMPDIR/out.bam\n",
    "rm $TMPDIR/out.bam\n",
    "samtools index $TMPDIR/BSF_G_merge.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#copy the library files to use as control\n",
    "!cp BSF_01/res4/BSF_01/BSF_01_sorted.bam $TMPDIR/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create a folder to store macs result\n",
    "!mkdir -p merge_macs_BSF_G\n",
    "\n",
    "#use macs2 to call broad peaks\n",
    "!macs2 callpeak -t $TMPDIR/BSF_G_merge.bam -f BAMPE \\\n",
    "-c $TMPDIR/BSF_01_sorted.bam \\\n",
    "--min-length 300 --max-gap 1 --broad --nomodel --keep-dup all\\\n",
    "-q 1 --broad-cutoff 1 --broad --llocal 3000 --outdir 'merge_macs_BSF_G'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load peak results and make a bed files\n",
    "#set coverage to one for visualization\n",
    "import pandas as pd\n",
    "indf = pd.read_csv('merge_macs_BSF_G/NA_peaks.xls',sep='\\t',comment='#')\n",
    "#GeneID\tChr\tStart\tEnd\tStrand\n",
    "indf = indf[list(indf.columns[0:3])]\n",
    "indf['coverage']=1\n",
    "#indf.columns = ['GeneID','Chr','Start','End']\n",
    "indf.to_csv('merge_macs_BSF_G/merge_macs_BSF_G.bed',sep='\\t',index=False,header=False)\n",
    "indf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make SAF file to get read counts\n",
    "indf = pd.read_csv('merge_macs_BSF_G/NA_peaks.xls',sep='\\t',comment='#')\n",
    "#GeneID\tChr\tStart\tEnd\tStrand\n",
    "indf = indf[[indf.columns[-1]] + list(indf.columns[0:3])]\n",
    "indf.columns = ['GeneID','Chr','Start','End']\n",
    "indf['Strand']='.'\n",
    "indf.to_csv('merge_macs_BSF_G/merge_macs_BSF_G.SAF',sep='\\t',index=False)\n",
    "indf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!less merge_macs_BSF_G/merge_macs_BSF_G.SAF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!featureCounts -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#we now count the read pairs for each B samples and barcoded B samples\n",
    "!featureCounts -p -B -C -M -O -T 8 -F SAF -a 'merge_macs_BSF_G/merge_macs_BSF_G.SAF' \\\n",
    "-o 'merge_macs_BSF_G/merge_macs_BSF_G_counts.txt' \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_01/res4/BSF_01/BSF_01_sorted.bam \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_06/res4/BSF_06/BSF_06_sorted.bam \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_07/res4/BSF_07/BSF_07_sorted.bam \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_08/res4/BSF_08/BSF_08_sorted.bam \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_01/res4/BSF_01/BSF_01_sorted_F.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_01/res4/BSF_01/BSF_01_sorted_R.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_01/res4/BSF_01/BSF_01_sorted_FR.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_01/res4/BSF_01/BSF_01_sorted_RR.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_06/res4/BSF_06/BSF_06_sorted_F.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_06/res4/BSF_06/BSF_06_sorted_R.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_06/res4/BSF_06/BSF_06_sorted_FR.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_06/res4/BSF_06/BSF_06_sorted_RR.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_07/res4/BSF_07/BSF_07_sorted_F.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_07/res4/BSF_07/BSF_07_sorted_R.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_07/res4/BSF_07/BSF_07_sorted_FR.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_07/res4/BSF_07/BSF_07_sorted_RR.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_08/res4/BSF_08/BSF_08_sorted_F.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_08/res4/BSF_08/BSF_08_sorted_R.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_08/res4/BSF_08/BSF_08_sorted_FR.bam  \\\n",
    "/cluster/majf_lab/mtinti/UTR/BSF_08/res4/BSF_08/BSF_08_sorted_RR.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>BSF_01</th>\n",
       "      <th>BSF_06</th>\n",
       "      <th>BSF_07</th>\n",
       "      <th>BSF_08</th>\n",
       "      <th>BSF_01_F</th>\n",
       "      <th>...</th>\n",
       "      <th>BSF_06_FR</th>\n",
       "      <th>BSF_06_RR</th>\n",
       "      <th>BSF_07_F</th>\n",
       "      <th>BSF_07_R</th>\n",
       "      <th>BSF_07_FR</th>\n",
       "      <th>BSF_07_RR</th>\n",
       "      <th>BSF_08_F</th>\n",
       "      <th>BSF_08_R</th>\n",
       "      <th>BSF_08_FR</th>\n",
       "      <th>BSF_08_RR</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NA_peak_1</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>248</td>\n",
       "      <td>660</td>\n",
       "      <td>.</td>\n",
       "      <td>413</td>\n",
       "      <td>207</td>\n",
       "      <td>455</td>\n",
       "      <td>363</td>\n",
       "      <td>310</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_2</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>770</td>\n",
       "      <td>1914</td>\n",
       "      <td>.</td>\n",
       "      <td>1145</td>\n",
       "      <td>1117</td>\n",
       "      <td>4607</td>\n",
       "      <td>5774</td>\n",
       "      <td>3073</td>\n",
       "      <td>29</td>\n",
       "      <td>...</td>\n",
       "      <td>25</td>\n",
       "      <td>92</td>\n",
       "      <td>112</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>135</td>\n",
       "      <td>58</td>\n",
       "      <td>8</td>\n",
       "      <td>22</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_3</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>27809</td>\n",
       "      <td>28704</td>\n",
       "      <td>.</td>\n",
       "      <td>896</td>\n",
       "      <td>22</td>\n",
       "      <td>122</td>\n",
       "      <td>74</td>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_4</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>43681</td>\n",
       "      <td>44162</td>\n",
       "      <td>.</td>\n",
       "      <td>482</td>\n",
       "      <td>269</td>\n",
       "      <td>888</td>\n",
       "      <td>749</td>\n",
       "      <td>1165</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>45</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>35</td>\n",
       "      <td>19</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>71</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_5</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>50638</td>\n",
       "      <td>50964</td>\n",
       "      <td>.</td>\n",
       "      <td>327</td>\n",
       "      <td>107</td>\n",
       "      <td>143</td>\n",
       "      <td>113</td>\n",
       "      <td>138</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Chr  Start    End Strand  Length  BSF_01  BSF_06  BSF_07  \\\n",
       "Geneid                                                                    \n",
       "NA_peak_1  11L3_v3    248    660      .     413     207     455     363   \n",
       "NA_peak_2  11L3_v3    770   1914      .    1145    1117    4607    5774   \n",
       "NA_peak_3  11L3_v3  27809  28704      .     896      22     122      74   \n",
       "NA_peak_4  11L3_v3  43681  44162      .     482     269     888     749   \n",
       "NA_peak_5  11L3_v3  50638  50964      .     327     107     143     113   \n",
       "\n",
       "           BSF_08  BSF_01_F  ...  BSF_06_FR  BSF_06_RR  BSF_07_F  BSF_07_R  \\\n",
       "Geneid                       ...                                             \n",
       "NA_peak_1     310         0  ...         17          1         0         3   \n",
       "NA_peak_2    3073        29  ...         25         92       112         2   \n",
       "NA_peak_3      72         0  ...          1          5         2         0   \n",
       "NA_peak_4    1165         0  ...         45         20         0         0   \n",
       "NA_peak_5     138         0  ...          0          0         0         0   \n",
       "\n",
       "           BSF_07_FR  BSF_07_RR  BSF_08_F  BSF_08_R  BSF_08_FR  BSF_08_RR  \n",
       "Geneid                                                                     \n",
       "NA_peak_1         10          1         0         0         16          2  \n",
       "NA_peak_2         16        135        58         8         22         83  \n",
       "NA_peak_3          0          1         6         0          1          5  \n",
       "NA_peak_4         35         19         3         0         71         25  \n",
       "NA_peak_5          0          0         0         0          0          0  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#get the counts of all samples\n",
    "import pandas as pd\n",
    "df = pd.read_csv('merge_macs_BSF_G/merge_macs_BSF_G_counts.txt',comment='#',sep='\\t',index_col=[0])\n",
    "counts_col = list(df.columns[5:])\n",
    "new_names = [n.split('/')[-1].split('.')[0].replace('_sorted','') for n in counts_col]\n",
    "new_names = dict(zip(counts_col,new_names))\n",
    "df = df.rename(new_names,axis=1)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NA_peak_1</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>248</td>\n",
       "      <td>660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_2</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>770</td>\n",
       "      <td>1914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_3</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>27809</td>\n",
       "      <td>28704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_4</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>43681</td>\n",
       "      <td>44162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_5</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>50638</td>\n",
       "      <td>50964</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Chr  Start    End\n",
       "Geneid                          \n",
       "NA_peak_1  11L3_v3    248    660\n",
       "NA_peak_2  11L3_v3    770   1914\n",
       "NA_peak_3  11L3_v3  27809  28704\n",
       "NA_peak_4  11L3_v3  43681  44162\n",
       "NA_peak_5  11L3_v3  50638  50964"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Prepare a final table\n",
    "final_table=pd.DataFrame(index=df.index)\n",
    "final_table['Chr']=df.Chr\n",
    "final_table['Start']=df.Start\n",
    "final_table['End']=df.End\n",
    "final_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Control_total</th>\n",
       "      <th>Control_F</th>\n",
       "      <th>Control_R</th>\n",
       "      <th>Treatment_total</th>\n",
       "      <th>Treatment_F</th>\n",
       "      <th>Treatment_R</th>\n",
       "      <th>Treatment_fraction_F</th>\n",
       "      <th>Treatment_fraction_R</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NA_peak_1</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>248</td>\n",
       "      <td>660</td>\n",
       "      <td>207</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>1128</td>\n",
       "      <td>4</td>\n",
       "      <td>48</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.923077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_2</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>770</td>\n",
       "      <td>1914</td>\n",
       "      <td>1117</td>\n",
       "      <td>58</td>\n",
       "      <td>20</td>\n",
       "      <td>13454</td>\n",
       "      <td>575</td>\n",
       "      <td>78</td>\n",
       "      <td>0.880551</td>\n",
       "      <td>0.119449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_3</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>27809</td>\n",
       "      <td>28704</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>268</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>0.925926</td>\n",
       "      <td>0.074074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_4</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>43681</td>\n",
       "      <td>44162</td>\n",
       "      <td>269</td>\n",
       "      <td>25</td>\n",
       "      <td>4</td>\n",
       "      <td>2802</td>\n",
       "      <td>67</td>\n",
       "      <td>151</td>\n",
       "      <td>0.307339</td>\n",
       "      <td>0.692661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_5</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>50638</td>\n",
       "      <td>50964</td>\n",
       "      <td>107</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>394</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Chr  Start    End  Control_total  Control_F  Control_R  \\\n",
       "Geneid                                                                  \n",
       "NA_peak_1  11L3_v3    248    660            207          3          7   \n",
       "NA_peak_2  11L3_v3    770   1914           1117         58         20   \n",
       "NA_peak_3  11L3_v3  27809  28704             22          0          1   \n",
       "NA_peak_4  11L3_v3  43681  44162            269         25          4   \n",
       "NA_peak_5  11L3_v3  50638  50964            107          0          0   \n",
       "\n",
       "           Treatment_total  Treatment_F  Treatment_R  Treatment_fraction_F  \\\n",
       "Geneid                                                                       \n",
       "NA_peak_1             1128            4           48              0.076923   \n",
       "NA_peak_2            13454          575           78              0.880551   \n",
       "NA_peak_3              268           25            2              0.925926   \n",
       "NA_peak_4             2802           67          151              0.307339   \n",
       "NA_peak_5              394            0            0                   NaN   \n",
       "\n",
       "           Treatment_fraction_R  \n",
       "Geneid                           \n",
       "NA_peak_1              0.923077  \n",
       "NA_peak_2              0.119449  \n",
       "NA_peak_3              0.074074  \n",
       "NA_peak_4              0.692661  \n",
       "NA_peak_5                   NaN  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#prepare the control\n",
    "final_table['Control_total']=df['BSF_01']\n",
    "final_table['Control_F']=df['BSF_01_F']+df['BSF_01_RR']\n",
    "final_table['Control_R']=df['BSF_01_R']+df['BSF_01_FR']\n",
    "#prepare treatment\n",
    "final_table['Treatment_total']=df['BSF_06']+df['BSF_07']+df['BSF_08']\n",
    "\n",
    "final_table['Treatment_F']=(df['BSF_06_F']+df['BSF_07_F']+df['BSF_08_F']+\n",
    "                            df['BSF_06_RR']+df['BSF_07_RR']+df['BSF_08_RR'])\n",
    "\n",
    "final_table['Treatment_R']=(df['BSF_06_R']+df['BSF_07_R']+df['BSF_08_R']+\n",
    "                            df['BSF_06_FR']+df['BSF_07_FR']+df['BSF_08_FR'])\n",
    "\n",
    "#fraction of reads in forward oritentation\n",
    "final_table['Treatment_fraction_F']=final_table['Treatment_F']/(final_table['Treatment_F']+final_table['Treatment_R'])\n",
    "#by definition\n",
    "final_table['Treatment_fraction_R']=1-final_table['Treatment_fraction_F']\n",
    "final_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Control_total</th>\n",
       "      <th>Control_F</th>\n",
       "      <th>Control_R</th>\n",
       "      <th>Treatment_total</th>\n",
       "      <th>Treatment_F</th>\n",
       "      <th>Treatment_R</th>\n",
       "      <th>Treatment_fraction_F</th>\n",
       "      <th>Treatment_fraction_R</th>\n",
       "      <th>Treatment_total_corrected_F</th>\n",
       "      <th>Treatment_total_corrected_R</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NA_peak_1</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>248</td>\n",
       "      <td>660</td>\n",
       "      <td>207</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>1128</td>\n",
       "      <td>4</td>\n",
       "      <td>48</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>86.769231</td>\n",
       "      <td>1041.230769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_2</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>770</td>\n",
       "      <td>1914</td>\n",
       "      <td>1117</td>\n",
       "      <td>58</td>\n",
       "      <td>20</td>\n",
       "      <td>13454</td>\n",
       "      <td>575</td>\n",
       "      <td>78</td>\n",
       "      <td>0.880551</td>\n",
       "      <td>0.119449</td>\n",
       "      <td>11846.937213</td>\n",
       "      <td>1607.062787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_3</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>27809</td>\n",
       "      <td>28704</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>268</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>0.925926</td>\n",
       "      <td>0.074074</td>\n",
       "      <td>248.148148</td>\n",
       "      <td>19.851852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_4</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>43681</td>\n",
       "      <td>44162</td>\n",
       "      <td>269</td>\n",
       "      <td>25</td>\n",
       "      <td>4</td>\n",
       "      <td>2802</td>\n",
       "      <td>67</td>\n",
       "      <td>151</td>\n",
       "      <td>0.307339</td>\n",
       "      <td>0.692661</td>\n",
       "      <td>861.165138</td>\n",
       "      <td>1940.834862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NA_peak_5</th>\n",
       "      <td>11L3_v3</td>\n",
       "      <td>50638</td>\n",
       "      <td>50964</td>\n",
       "      <td>107</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>394</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Chr  Start    End  Control_total  Control_F  Control_R  \\\n",
       "Geneid                                                                  \n",
       "NA_peak_1  11L3_v3    248    660            207          3          7   \n",
       "NA_peak_2  11L3_v3    770   1914           1117         58         20   \n",
       "NA_peak_3  11L3_v3  27809  28704             22          0          1   \n",
       "NA_peak_4  11L3_v3  43681  44162            269         25          4   \n",
       "NA_peak_5  11L3_v3  50638  50964            107          0          0   \n",
       "\n",
       "           Treatment_total  Treatment_F  Treatment_R  Treatment_fraction_F  \\\n",
       "Geneid                                                                       \n",
       "NA_peak_1             1128            4           48              0.076923   \n",
       "NA_peak_2            13454          575           78              0.880551   \n",
       "NA_peak_3              268           25            2              0.925926   \n",
       "NA_peak_4             2802           67          151              0.307339   \n",
       "NA_peak_5              394            0            0                   NaN   \n",
       "\n",
       "           Treatment_fraction_R  Treatment_total_corrected_F  \\\n",
       "Geneid                                                         \n",
       "NA_peak_1              0.923077                    86.769231   \n",
       "NA_peak_2              0.119449                 11846.937213   \n",
       "NA_peak_3              0.074074                   248.148148   \n",
       "NA_peak_4              0.692661                   861.165138   \n",
       "NA_peak_5                   NaN                          NaN   \n",
       "\n",
       "           Treatment_total_corrected_R  \n",
       "Geneid                                  \n",
       "NA_peak_1                  1041.230769  \n",
       "NA_peak_2                  1607.062787  \n",
       "NA_peak_3                    19.851852  \n",
       "NA_peak_4                  1940.834862  \n",
       "NA_peak_5                          NaN  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_table['Treatment_total_corrected_F']=final_table['Treatment_total']*final_table['Treatment_fraction_F']\n",
    "final_table['Treatment_total_corrected_R']=final_table['Treatment_total']*final_table['Treatment_fraction_R']\n",
    "final_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11L3_v3\t248\t660\tNA_peak_1\n",
      "11L3_v3\t770\t1914\tNA_peak_2\n",
      "11L3_v3\t27809\t28704\tNA_peak_3\n",
      "11L3_v3\t43681\t44162\tNA_peak_4\n",
      "11L3_v3\t50638\t50964\tNA_peak_5\n",
      "11L3_v3\t50975\t51514\tNA_peak_6\n",
      "11L3_v3\t52360\t53489\tNA_peak_7\n",
      "11L3_v3\t55134\t56732\tNA_peak_8\n",
      "11L3_v3\t58961\t59515\tNA_peak_9\n",
      "11L3_v3\t59964\t60927\tNA_peak_10\n"
     ]
    }
   ],
   "source": [
    "#create a bed file\n",
    "bed_file = df[['Chr', 'Start', 'End']]\n",
    "bed_file=bed_file.reset_index()\n",
    "bed_file = bed_file[['Chr', 'Start', 'End', 'Geneid']]\n",
    "bed_file=bed_file.rename({'Geneid':'Peak_Id'},axis=1)\n",
    "bed_file[['Chr', 'Start', 'End', 'Peak_Id']].to_csv('gla_peaks_final.bed', sep='\\t',\n",
    "                                                   header=False, index=False)\n",
    "!head 'gla_peaks_final.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(284, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>Tb927.9.15950</td>\n",
       "      <td>UTR_contained</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>Tb927.9.6570</td>\n",
       "      <td>UTR_contained</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>Tb927.9.6830</td>\n",
       "      <td>UTR_contained</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>Tb927.9.7500</td>\n",
       "      <td>UTR_contained</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>Tb927.9.7850</td>\n",
       "      <td>UTR_contained</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene         reason\n",
       "279  Tb927.9.15950  UTR_contained\n",
       "280   Tb927.9.6570  UTR_contained\n",
       "281   Tb927.9.6830  UTR_contained\n",
       "282   Tb927.9.7500  UTR_contained\n",
       "283   Tb927.9.7850  UTR_contained"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bl =pd.read_csv('blacklisted gene.csv',engine='python',header=None)\n",
    "bl.columns = ['gene','reason']\n",
    "bl['gene']=[n.strip() for n in bl['gene']]\n",
    "print(bl.shape)\n",
    "bl.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "bedtools intersect -a <(bedtools sort -i gla_peaks_final.bed) \\\n",
    "-b <(bedtools sort -i UTR3_final.bed) -wao > gla_peak_belongs_to_3UTR.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#use awk to select only the fully contained\n",
    "!awk '$(NF-2) != \"-1\"' gla_peak_belongs_to_3UTR.bed > gla_peak_belongs_to_3UTR_filtered.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('gla_peak_belongs_to_3UTR_filtered.bed',sep='\\t',header=None)\n",
    "df.columns = ['Chromosome','Peak_Start','Peak_End','Peak_Name',\n",
    "              'UTR_Chromosome','UTR_Start','UTR_End','UTR_Gene_ID','UTR_Score','UTR_Strand','UTR_Peak_overlap']\n",
    "del df['UTR_Score']\n",
    "del df['UTR_Chromosome']\n",
    "#df.head()\n",
    "\n",
    "df_plus = df[df['UTR_Strand']=='+']\n",
    "df_minus = df[df['UTR_Strand']=='-']\n",
    "df_plus = df_plus.drop_duplicates(subset=['Peak_Name'],keep='first')\n",
    "df_minus = df_minus.drop_duplicates(subset=['Peak_Name'],keep='last')\n",
    "df = pd.concat([df_plus,df_minus])\n",
    "df=df.sort_values(by=['Chromosome','Peak_Start','Peak_End'])\n",
    "#df[df['Peak_Name']=='NA_peak_1371']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2212, 9)\n",
      "(2212, 19)\n"
     ]
    }
   ],
   "source": [
    "print(df.shape)\n",
    "df = df.merge(final_table.iloc[:,3:],left_on='Peak_Name',right_index=True,how='left')\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Peak_Start</th>\n",
       "      <th>Peak_End</th>\n",
       "      <th>Peak_Name</th>\n",
       "      <th>UTR_Start</th>\n",
       "      <th>UTR_End</th>\n",
       "      <th>UTR_Gene_ID</th>\n",
       "      <th>UTR_Strand</th>\n",
       "      <th>UTR_Peak_overlap</th>\n",
       "      <th>Control_total</th>\n",
       "      <th>Control_F</th>\n",
       "      <th>Control_R</th>\n",
       "      <th>Treatment_total</th>\n",
       "      <th>Treatment_F</th>\n",
       "      <th>Treatment_R</th>\n",
       "      <th>Treatment_fraction_F</th>\n",
       "      <th>Treatment_fraction_R</th>\n",
       "      <th>Treatment_total_corrected_F</th>\n",
       "      <th>Treatment_total_corrected_R</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>60141</td>\n",
       "      <td>61723</td>\n",
       "      <td>NA_peak_188</td>\n",
       "      <td>59853</td>\n",
       "      <td>60265</td>\n",
       "      <td>Tb927.1.120</td>\n",
       "      <td>-</td>\n",
       "      <td>124</td>\n",
       "      <td>4881</td>\n",
       "      <td>201</td>\n",
       "      <td>351</td>\n",
       "      <td>32261</td>\n",
       "      <td>1561</td>\n",
       "      <td>576</td>\n",
       "      <td>0.730463</td>\n",
       "      <td>0.269537</td>\n",
       "      <td>23565.475433</td>\n",
       "      <td>8695.524567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>112710</td>\n",
       "      <td>113415</td>\n",
       "      <td>NA_peak_211</td>\n",
       "      <td>112744</td>\n",
       "      <td>112971</td>\n",
       "      <td>Tb927.1.280</td>\n",
       "      <td>+</td>\n",
       "      <td>227</td>\n",
       "      <td>2827</td>\n",
       "      <td>35</td>\n",
       "      <td>17</td>\n",
       "      <td>10330</td>\n",
       "      <td>164</td>\n",
       "      <td>35</td>\n",
       "      <td>0.824121</td>\n",
       "      <td>0.175879</td>\n",
       "      <td>8513.165829</td>\n",
       "      <td>1816.834171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>144650</td>\n",
       "      <td>144961</td>\n",
       "      <td>NA_peak_228</td>\n",
       "      <td>144595</td>\n",
       "      <td>144725</td>\n",
       "      <td>Tb927.1.380</td>\n",
       "      <td>-</td>\n",
       "      <td>75</td>\n",
       "      <td>1985</td>\n",
       "      <td>45</td>\n",
       "      <td>23</td>\n",
       "      <td>6948</td>\n",
       "      <td>28</td>\n",
       "      <td>185</td>\n",
       "      <td>0.131455</td>\n",
       "      <td>0.868545</td>\n",
       "      <td>913.352113</td>\n",
       "      <td>6034.647887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>149354</td>\n",
       "      <td>149874</td>\n",
       "      <td>NA_peak_231</td>\n",
       "      <td>148205</td>\n",
       "      <td>151108</td>\n",
       "      <td>Tb927.1.400</td>\n",
       "      <td>+</td>\n",
       "      <td>520</td>\n",
       "      <td>915</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>4968</td>\n",
       "      <td>40</td>\n",
       "      <td>35</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>2649.600000</td>\n",
       "      <td>2318.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>150298</td>\n",
       "      <td>150742</td>\n",
       "      <td>NA_peak_232</td>\n",
       "      <td>148205</td>\n",
       "      <td>151108</td>\n",
       "      <td>Tb927.1.400</td>\n",
       "      <td>+</td>\n",
       "      <td>444</td>\n",
       "      <td>1241</td>\n",
       "      <td>70</td>\n",
       "      <td>34</td>\n",
       "      <td>5080</td>\n",
       "      <td>137</td>\n",
       "      <td>135</td>\n",
       "      <td>0.503676</td>\n",
       "      <td>0.496324</td>\n",
       "      <td>2558.676471</td>\n",
       "      <td>2521.323529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>186068</td>\n",
       "      <td>188200</td>\n",
       "      <td>NA_peak_255</td>\n",
       "      <td>186030</td>\n",
       "      <td>187218</td>\n",
       "      <td>Tb927.1.480</td>\n",
       "      <td>+</td>\n",
       "      <td>1150</td>\n",
       "      <td>803</td>\n",
       "      <td>35</td>\n",
       "      <td>6</td>\n",
       "      <td>6706</td>\n",
       "      <td>39</td>\n",
       "      <td>54</td>\n",
       "      <td>0.419355</td>\n",
       "      <td>0.580645</td>\n",
       "      <td>2812.193548</td>\n",
       "      <td>3893.806452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>190391</td>\n",
       "      <td>191332</td>\n",
       "      <td>NA_peak_257</td>\n",
       "      <td>190086</td>\n",
       "      <td>190499</td>\n",
       "      <td>Tb927.1.490</td>\n",
       "      <td>+</td>\n",
       "      <td>108</td>\n",
       "      <td>4160</td>\n",
       "      <td>14</td>\n",
       "      <td>109</td>\n",
       "      <td>20332</td>\n",
       "      <td>52</td>\n",
       "      <td>526</td>\n",
       "      <td>0.089965</td>\n",
       "      <td>0.910035</td>\n",
       "      <td>1829.176471</td>\n",
       "      <td>18502.823529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>212303</td>\n",
       "      <td>213619</td>\n",
       "      <td>NA_peak_262</td>\n",
       "      <td>211572</td>\n",
       "      <td>212454</td>\n",
       "      <td>Tb927.1.580</td>\n",
       "      <td>-</td>\n",
       "      <td>151</td>\n",
       "      <td>763</td>\n",
       "      <td>79</td>\n",
       "      <td>62</td>\n",
       "      <td>9415</td>\n",
       "      <td>121</td>\n",
       "      <td>640</td>\n",
       "      <td>0.159001</td>\n",
       "      <td>0.840999</td>\n",
       "      <td>1496.997372</td>\n",
       "      <td>7918.002628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>216015</td>\n",
       "      <td>217313</td>\n",
       "      <td>NA_peak_263</td>\n",
       "      <td>214185</td>\n",
       "      <td>216163</td>\n",
       "      <td>Tb927.1.600</td>\n",
       "      <td>-</td>\n",
       "      <td>148</td>\n",
       "      <td>707</td>\n",
       "      <td>8</td>\n",
       "      <td>40</td>\n",
       "      <td>7578</td>\n",
       "      <td>78</td>\n",
       "      <td>271</td>\n",
       "      <td>0.223496</td>\n",
       "      <td>0.776504</td>\n",
       "      <td>1693.650430</td>\n",
       "      <td>5884.349570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>218980</td>\n",
       "      <td>219871</td>\n",
       "      <td>NA_peak_264</td>\n",
       "      <td>219177</td>\n",
       "      <td>219567</td>\n",
       "      <td>Tb927.1.630</td>\n",
       "      <td>-</td>\n",
       "      <td>390</td>\n",
       "      <td>110</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>802</td>\n",
       "      <td>13</td>\n",
       "      <td>50</td>\n",
       "      <td>0.206349</td>\n",
       "      <td>0.793651</td>\n",
       "      <td>165.492063</td>\n",
       "      <td>636.507937</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome  Peak_Start  Peak_End    Peak_Name  UTR_Start  UTR_End  \\\n",
       "0  Tb927_01_v5.1       60141     61723  NA_peak_188      59853    60265   \n",
       "1  Tb927_01_v5.1      112710    113415  NA_peak_211     112744   112971   \n",
       "2  Tb927_01_v5.1      144650    144961  NA_peak_228     144595   144725   \n",
       "3  Tb927_01_v5.1      149354    149874  NA_peak_231     148205   151108   \n",
       "4  Tb927_01_v5.1      150298    150742  NA_peak_232     148205   151108   \n",
       "5  Tb927_01_v5.1      186068    188200  NA_peak_255     186030   187218   \n",
       "6  Tb927_01_v5.1      190391    191332  NA_peak_257     190086   190499   \n",
       "7  Tb927_01_v5.1      212303    213619  NA_peak_262     211572   212454   \n",
       "8  Tb927_01_v5.1      216015    217313  NA_peak_263     214185   216163   \n",
       "9  Tb927_01_v5.1      218980    219871  NA_peak_264     219177   219567   \n",
       "\n",
       "   UTR_Gene_ID UTR_Strand  UTR_Peak_overlap  Control_total  Control_F  \\\n",
       "0  Tb927.1.120          -               124           4881        201   \n",
       "1  Tb927.1.280          +               227           2827         35   \n",
       "2  Tb927.1.380          -                75           1985         45   \n",
       "3  Tb927.1.400          +               520            915         11   \n",
       "4  Tb927.1.400          +               444           1241         70   \n",
       "5  Tb927.1.480          +              1150            803         35   \n",
       "6  Tb927.1.490          +               108           4160         14   \n",
       "7  Tb927.1.580          -               151            763         79   \n",
       "8  Tb927.1.600          -               148            707          8   \n",
       "9  Tb927.1.630          -               390            110          1   \n",
       "\n",
       "   Control_R  Treatment_total  Treatment_F  Treatment_R  Treatment_fraction_F  \\\n",
       "0        351            32261         1561          576              0.730463   \n",
       "1         17            10330          164           35              0.824121   \n",
       "2         23             6948           28          185              0.131455   \n",
       "3         13             4968           40           35              0.533333   \n",
       "4         34             5080          137          135              0.503676   \n",
       "5          6             6706           39           54              0.419355   \n",
       "6        109            20332           52          526              0.089965   \n",
       "7         62             9415          121          640              0.159001   \n",
       "8         40             7578           78          271              0.223496   \n",
       "9          8              802           13           50              0.206349   \n",
       "\n",
       "   Treatment_fraction_R  Treatment_total_corrected_F  \\\n",
       "0              0.269537                 23565.475433   \n",
       "1              0.175879                  8513.165829   \n",
       "2              0.868545                   913.352113   \n",
       "3              0.466667                  2649.600000   \n",
       "4              0.496324                  2558.676471   \n",
       "5              0.580645                  2812.193548   \n",
       "6              0.910035                  1829.176471   \n",
       "7              0.840999                  1496.997372   \n",
       "8              0.776504                  1693.650430   \n",
       "9              0.793651                   165.492063   \n",
       "\n",
       "   Treatment_total_corrected_R  \n",
       "0                  8695.524567  \n",
       "1                  1816.834171  \n",
       "2                  6034.647887  \n",
       "3                  2318.400000  \n",
       "4                  2521.323529  \n",
       "5                  3893.806452  \n",
       "6                 18502.823529  \n",
       "7                  7918.002628  \n",
       "8                  5884.349570  \n",
       "9                   636.507937  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_download(X):\n",
    "    if X['UTR_Strand']=='-':\n",
    "        orient = 'r'\n",
    "    else:\n",
    "        orient = 'f'\n",
    "    chro =  X['Chromosome']  \n",
    "    p_start =  int(X['Peak_Start'])\n",
    "    p_end =  int(X['Peak_End'])\n",
    "    u_start =  int(X['UTR_Start'])\n",
    "    u_end =  int(X['UTR_End'])  \n",
    "    \n",
    "    start = max(p_start,u_start)\n",
    "    end = min(p_end,u_end)\n",
    "    download_string = f'{chro}:{start}..{end}:{orient}'\n",
    "    return download_string           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['download_seq']=df.apply(format_download,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_fully_contained(X):\n",
    "    p_start =  int(X['Peak_Start'])\n",
    "    p_end =  int(X['Peak_End'])\n",
    "    u_start =  int(X['UTR_Start'])\n",
    "    u_end =  int(X['UTR_End'])\n",
    "    if (p_start>=u_start) and (p_end<=u_end):\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "df['is_fully_contained'] = df.apply(is_fully_contained,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(260, 21)"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['is_fully_contained']].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2212, 21)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Peak_Start</th>\n",
       "      <th>Peak_End</th>\n",
       "      <th>Peak_Name</th>\n",
       "      <th>UTR_Start</th>\n",
       "      <th>UTR_End</th>\n",
       "      <th>UTR_Gene_ID</th>\n",
       "      <th>UTR_Strand</th>\n",
       "      <th>UTR_Peak_overlap</th>\n",
       "      <th>Control_total</th>\n",
       "      <th>...</th>\n",
       "      <th>Control_R</th>\n",
       "      <th>Treatment_total</th>\n",
       "      <th>Treatment_F</th>\n",
       "      <th>Treatment_R</th>\n",
       "      <th>Treatment_fraction_F</th>\n",
       "      <th>Treatment_fraction_R</th>\n",
       "      <th>Treatment_total_corrected_F</th>\n",
       "      <th>Treatment_total_corrected_R</th>\n",
       "      <th>download_seq</th>\n",
       "      <th>is_fully_contained</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>60141</td>\n",
       "      <td>61723</td>\n",
       "      <td>NA_peak_188</td>\n",
       "      <td>59853</td>\n",
       "      <td>60265</td>\n",
       "      <td>Tb927.1.120</td>\n",
       "      <td>-</td>\n",
       "      <td>124</td>\n",
       "      <td>4881</td>\n",
       "      <td>...</td>\n",
       "      <td>351</td>\n",
       "      <td>32261</td>\n",
       "      <td>1561</td>\n",
       "      <td>576</td>\n",
       "      <td>0.730463</td>\n",
       "      <td>0.269537</td>\n",
       "      <td>23565.475433</td>\n",
       "      <td>8695.524567</td>\n",
       "      <td>Tb927_01_v5.1:60141..60265:r</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>112710</td>\n",
       "      <td>113415</td>\n",
       "      <td>NA_peak_211</td>\n",
       "      <td>112744</td>\n",
       "      <td>112971</td>\n",
       "      <td>Tb927.1.280</td>\n",
       "      <td>+</td>\n",
       "      <td>227</td>\n",
       "      <td>2827</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>10330</td>\n",
       "      <td>164</td>\n",
       "      <td>35</td>\n",
       "      <td>0.824121</td>\n",
       "      <td>0.175879</td>\n",
       "      <td>8513.165829</td>\n",
       "      <td>1816.834171</td>\n",
       "      <td>Tb927_01_v5.1:112744..112971:f</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>144650</td>\n",
       "      <td>144961</td>\n",
       "      <td>NA_peak_228</td>\n",
       "      <td>144595</td>\n",
       "      <td>144725</td>\n",
       "      <td>Tb927.1.380</td>\n",
       "      <td>-</td>\n",
       "      <td>75</td>\n",
       "      <td>1985</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>6948</td>\n",
       "      <td>28</td>\n",
       "      <td>185</td>\n",
       "      <td>0.131455</td>\n",
       "      <td>0.868545</td>\n",
       "      <td>913.352113</td>\n",
       "      <td>6034.647887</td>\n",
       "      <td>Tb927_01_v5.1:144650..144725:r</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>149354</td>\n",
       "      <td>149874</td>\n",
       "      <td>NA_peak_231</td>\n",
       "      <td>148205</td>\n",
       "      <td>151108</td>\n",
       "      <td>Tb927.1.400</td>\n",
       "      <td>+</td>\n",
       "      <td>520</td>\n",
       "      <td>915</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>4968</td>\n",
       "      <td>40</td>\n",
       "      <td>35</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>2649.600000</td>\n",
       "      <td>2318.400000</td>\n",
       "      <td>Tb927_01_v5.1:149354..149874:f</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Tb927_01_v5.1</td>\n",
       "      <td>150298</td>\n",
       "      <td>150742</td>\n",
       "      <td>NA_peak_232</td>\n",
       "      <td>148205</td>\n",
       "      <td>151108</td>\n",
       "      <td>Tb927.1.400</td>\n",
       "      <td>+</td>\n",
       "      <td>444</td>\n",
       "      <td>1241</td>\n",
       "      <td>...</td>\n",
       "      <td>34</td>\n",
       "      <td>5080</td>\n",
       "      <td>137</td>\n",
       "      <td>135</td>\n",
       "      <td>0.503676</td>\n",
       "      <td>0.496324</td>\n",
       "      <td>2558.676471</td>\n",
       "      <td>2521.323529</td>\n",
       "      <td>Tb927_01_v5.1:150298..150742:f</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome  Peak_Start  Peak_End    Peak_Name  UTR_Start  UTR_End  \\\n",
       "0  Tb927_01_v5.1       60141     61723  NA_peak_188      59853    60265   \n",
       "1  Tb927_01_v5.1      112710    113415  NA_peak_211     112744   112971   \n",
       "2  Tb927_01_v5.1      144650    144961  NA_peak_228     144595   144725   \n",
       "3  Tb927_01_v5.1      149354    149874  NA_peak_231     148205   151108   \n",
       "4  Tb927_01_v5.1      150298    150742  NA_peak_232     148205   151108   \n",
       "\n",
       "   UTR_Gene_ID UTR_Strand  UTR_Peak_overlap  Control_total  ...  Control_R  \\\n",
       "0  Tb927.1.120          -               124           4881  ...        351   \n",
       "1  Tb927.1.280          +               227           2827  ...         17   \n",
       "2  Tb927.1.380          -                75           1985  ...         23   \n",
       "3  Tb927.1.400          +               520            915  ...         13   \n",
       "4  Tb927.1.400          +               444           1241  ...         34   \n",
       "\n",
       "   Treatment_total  Treatment_F  Treatment_R  Treatment_fraction_F  \\\n",
       "0            32261         1561          576              0.730463   \n",
       "1            10330          164           35              0.824121   \n",
       "2             6948           28          185              0.131455   \n",
       "3             4968           40           35              0.533333   \n",
       "4             5080          137          135              0.503676   \n",
       "\n",
       "   Treatment_fraction_R  Treatment_total_corrected_F  \\\n",
       "0              0.269537                 23565.475433   \n",
       "1              0.175879                  8513.165829   \n",
       "2              0.868545                   913.352113   \n",
       "3              0.466667                  2649.600000   \n",
       "4              0.496324                  2558.676471   \n",
       "\n",
       "   Treatment_total_corrected_R                    download_seq  \\\n",
       "0                  8695.524567    Tb927_01_v5.1:60141..60265:r   \n",
       "1                  1816.834171  Tb927_01_v5.1:112744..112971:f   \n",
       "2                  6034.647887  Tb927_01_v5.1:144650..144725:r   \n",
       "3                  2318.400000  Tb927_01_v5.1:149354..149874:f   \n",
       "4                  2521.323529  Tb927_01_v5.1:150298..150742:f   \n",
       "\n",
       "  is_fully_contained  \n",
       "0              False  \n",
       "1              False  \n",
       "2              False  \n",
       "3               True  \n",
       "4               True  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('Paper_Table_Ganciclovir.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sys==Python BuiltIn\n",
      "tqdm==4.64.0\n",
      "matplotlib==3.5.1\n",
      "numpy==1.21.6\n",
      "pandas==1.4.2\n",
      "types==unknown\n",
      "pkg_resources==unknown\n"
     ]
    }
   ],
   "source": [
    "# show versions of packages\n",
    "# adopted from https://stackoverflow.com/questions/40428931/package-for-listing-version-of-packages-used-in-a-jupyter-notebook\n",
    "import pkg_resources\n",
    "import types\n",
    "import sys\n",
    "def get_imports():\n",
    "    for name, val in globals().items():\n",
    "        if isinstance(val, types.ModuleType):\n",
    "            # Split ensures you get root package, \n",
    "            # not just imported function\n",
    "            name = val.__name__.split(\".\")[0]\n",
    "        elif isinstance(val, type):\n",
    "            name = val.__module__.split(\".\")[0]\n",
    "        # Some packages are weird and have different\n",
    "        # imported names vs. system/pip names. Unfortunately,\n",
    "        # there is no systematic way to get pip names from\n",
    "        # a package's imported name. You'll have to add\n",
    "        # exceptions to this list manually!\n",
    "        poorly_named_packages = {\n",
    "            \"sklearn\": \"scikit-learn\"\n",
    "        }\n",
    "        if name in poorly_named_packages.keys():\n",
    "            name = poorly_named_packages[name]\n",
    "        yield name.lower()\n",
    "imports = list(set(get_imports()))\n",
    "\n",
    "# The only way I found to get the version of the root package\n",
    "# from only the name of the package is to cross-check the names \n",
    "# of installed packages vs. imported packages\n",
    "modules = []\n",
    "for m in sys.builtin_module_names:\n",
    "    if m.lower() in imports and m !='builtins':\n",
    "        modules.append((m,'Python BuiltIn'))\n",
    "        imports.remove(m.lower())\n",
    "\n",
    "for m in pkg_resources.working_set:\n",
    "    if m.project_name.lower() in imports and m.project_name!=\"pip\":\n",
    "        modules.append((m.project_name, m.version))\n",
    "        imports.remove(m.project_name.lower())\n",
    "\n",
    "for m in sys.modules:\n",
    "    if m.lower() in imports and m !='builtins':\n",
    "        modules.append((m,'unknown'))\n",
    "\n",
    "# print('System=='+platform.system()+' '+platform.release()+'; Version=='+platform.version())\n",
    "for r in modules:\n",
    "    print(\"{}=={}\".format(*r))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#bedtools-2.31"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
