{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1cef22d3-467a-4f68-9b98-69e89546f981",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Tb927_01_v5.1',\n",
       " 'Tb927_02_v5.1',\n",
       " 'Tb927_03_v5.1',\n",
       " 'Tb927_04_v5.1',\n",
       " 'Tb927_05_v5.1',\n",
       " 'Tb927_06_v5.1',\n",
       " 'Tb927_07_v5.1',\n",
       " 'Tb927_08_v5.1',\n",
       " 'Tb927_09_v5.1',\n",
       " 'Tb927_10_v5.1',\n",
       " 'Tb927_10_v5.1']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromosome_list = [\n",
    "    f'Tb927_0{n}_v5.1' for n in range(1,10)\n",
    "]\n",
    "chromosome_list+=['Tb927_10_v5.1','Tb927_10_v5.1']\n",
    "chromosome_list=set(chromosome_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0beee62d-642d-4e06-aea6-719e0f9660aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "found 4514 3UTRs\n"
     ]
    }
   ],
   "source": [
    "c=0\n",
    "out = open('count_pca/count_pca_old.SAF','w') \n",
    "out.write('GeneID\\tChr\\tStart\\tEnd\\tStrand\\n')\n",
    "for l in open('/cluster/majf_lab/mtinti/UTR/TriTrypDB-46_TbruceiTREU927.gff'):\n",
    "    if 'three_prime_UTR' in l:\n",
    "        item_list = (l.split())\n",
    "        chromosome =item_list[0]\n",
    "        if chromosome in chromosome_list:\n",
    "            start =item_list[3]\n",
    "            end =item_list[4]\n",
    "            utr_id = item_list[8].split(';')[0].split('ID=utr_')[1]\n",
    "            out.write(f'{utr_id}\\t{chromosome}\\t{start}\\t{end}\\t.\\n')\n",
    "            c+=1\n",
    "print(f'found {c} 3UTRs')\n",
    "out.close()    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7271e9af-2149-4822-8e42-2d5be1956764",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GeneID\tChr\tStart\tEnd\tStrand\n",
      "Tb927.8.4730:mRNA_1\tTb927_08_v5.1\t1387684\t1389085\t.\n",
      "Tb927.7.5460:mRNA_2\tTb927_07_v5.1\t1475328\t1475778\t.\n",
      "Tb927.4.580:mRNA_1\tTb927_04_v5.1\t166931\t167193\t.\n",
      "Tb927.5.2260:mRNA_2\tTb927_05_v5.1\t762326\t762699\t.\n",
      "Tb927.4.740:mRNA_1\tTb927_04_v5.1\t205296\t205462\t.\n",
      "Tb927.9.3540:mRNA_2\tTb927_09_v5.1\t718521\t718658\t.\n",
      "Tb927.9.12440:mRNA_2\tTb927_09_v5.1\t1945610\t1945665\t.\n",
      "Tb927.8.5560:mRNA_2\tTb927_08_v5.1\t1640354\t1640521\t.\n",
      "Tb927.4.1040:mRNA_1\tTb927_04_v5.1\t279831\t280220\t.\n"
     ]
    }
   ],
   "source": [
    "!head 'count_pca/count_pca_old.SAF'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9552f78-6ce2-4307-8340-99d77c66e7c6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffc9783b-868c-48a7-85f2-c63695150efe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "dc170ec0-6c04-4ae7-af85-c62457155105",
   "metadata": {},
   "outputs": [],
   "source": [
    "#GeneID\tChr\tStart\tEnd\tStrand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "106dd957-0585-4c29-ace6-54cfa5bb11a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import pandas as pd\n",
    "#indf = pd.read_csv('UTR3_final.bed',sep='\\t',header=None)\n",
    "#indf.columns = ['Chr','Start','End','GeneID','Score','Strand']\n",
    "#indf['Strand']='.'\n",
    "#indf[['GeneID','Chr','Start','End','Strand']].to_csv('count_pca/count_pca.SAF',sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a63fe888-67fd-47fe-96e3-9bf29e815229",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v1.6.4\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m7 BAM files  \u001b[0m \u001b[0m                                   ||\n",
      "||                           \u001b[32mP\u001b[36m BSF_01_sorted.bam\u001b[0m \u001b[0m                             ||\n",
      "||                           \u001b[32mP\u001b[36m BSF_03_sorted.bam\u001b[0m \u001b[0m                             ||\n",
      "||                           \u001b[32mP\u001b[36m BSF_04_sorted.bam\u001b[0m \u001b[0m                             ||\n",
      "||                           \u001b[32mP\u001b[36m BSF_05_sorted.bam\u001b[0m \u001b[0m                             ||\n",
      "||                           \u001b[32mP\u001b[36m BSF_06_sorted.bam\u001b[0m \u001b[0m                             ||\n",
      "||                           \u001b[32mP\u001b[36m BSF_07_sorted.bam\u001b[0m \u001b[0m                             ||\n",
      "||                           \u001b[32mP\u001b[36m BSF_08_sorted.bam\u001b[0m \u001b[0m                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcount_pca_old.txt\u001b[0m \u001b[0m                               ||\n",
      "||                 Summary : \u001b[36mcount_pca_old.txt.summary\u001b[0m \u001b[0m                       ||\n",
      "||              Annotation : \u001b[36mcount_pca_old.SAF (SAF)\u001b[0m \u001b[0m                         ||\n",
      "||      Dir for temp files : \u001b[36mcount_pca\u001b[0m \u001b[0m                                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m16\u001b[0m \u001b[0m                                              ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||              Paired-end : \u001b[36myes\u001b[0m \u001b[0m                                             ||\n",
      "||      Multimapping reads : \u001b[36mcounted\u001b[0m \u001b[0m                                         ||\n",
      "|| Multi-overlapping reads : \u001b[36mcounted\u001b[0m \u001b[0m                                         ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||          Chimeric reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||        Both ends mapped : \u001b[36mrequired\u001b[0m \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file count_pca_old.SAF \u001b[0m... \u001b[0m                                ||\n",
      "||    Features : \u001b[36m4514\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m4514\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m10\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file BSF_01_sorted.bam... \u001b[0m                                     ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Assign alignments (paired-end) to features... \u001b[0m                          ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    WARNING: reads from the same pair were found not adjacent to each \u001b[0m      ||\n",
      "||             other in the input (due to read sorting by location or \u001b[0m        ||\n",
      "||             reporting of multi-mapping read pairs). \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    Pairing up the read pairs. \u001b[0m                                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    Total alignments : \u001b[36m28966126\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m3569171 (12.3%)\u001b[0m \u001b[0m                     ||\n",
      "||    Running time : \u001b[36m2.55 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file BSF_04_sorted.bam... \u001b[0m                                     ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Assign alignments (paired-end) to features... \u001b[0m                          ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    WARNING: reads from the same pair were found not adjacent to each \u001b[0m      ||\n",
      "||             other in the input (due to read sorting by location or \u001b[0m        ||\n",
      "||             reporting of multi-mapping read pairs). \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    Pairing up the read pairs. \u001b[0m                                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    Total alignments : \u001b[36m24322142\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m2763998 (11.4%)\u001b[0m \u001b[0m                     ||\n",
      "||    Running time : \u001b[36m2.32 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file BSF_05_sorted.bam... \u001b[0m                                     ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Assign alignments (paired-end) to features... \u001b[0m                          ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    WARNING: reads from the same pair were found not adjacent to each \u001b[0m      ||\n",
      "||             other in the input (due to read sorting by location or \u001b[0m        ||\n",
      "||             reporting of multi-mapping read pairs). \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    Pairing up the read pairs. \u001b[0m                                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    Total alignments : \u001b[36m28072278\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m2748830 (9.8%)\u001b[0m \u001b[0m                      ||\n",
      "||    Running time : \u001b[36m2.74 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file BSF_06_sorted.bam... \u001b[0m                                     ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Assign alignments (paired-end) to features... \u001b[0m                          ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    WARNING: reads from the same pair were found not adjacent to each \u001b[0m      ||\n",
      "||             other in the input (due to read sorting by location or \u001b[0m        ||\n",
      "||             reporting of multi-mapping read pairs). \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    Pairing up the read pairs. \u001b[0m                                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    Total alignments : \u001b[36m40862134\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m3790574 (9.3%)\u001b[0m \u001b[0m                      ||\n",
      "||    Running time : \u001b[36m3.03 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file BSF_07_sorted.bam... \u001b[0m                                     ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Assign alignments (paired-end) to features... \u001b[0m                          ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    WARNING: reads from the same pair were found not adjacent to each \u001b[0m      ||\n",
      "||             other in the input (due to read sorting by location or \u001b[0m        ||\n",
      "||             reporting of multi-mapping read pairs). \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    Pairing up the read pairs. \u001b[0m                                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    Total alignments : \u001b[36m32613864\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m2856394 (8.8%)\u001b[0m \u001b[0m                      ||\n",
      "||    Running time : \u001b[36m2.44 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file BSF_08_sorted.bam... \u001b[0m                                     ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Assign alignments (paired-end) to features... \u001b[0m                          ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    WARNING: reads from the same pair were found not adjacent to each \u001b[0m      ||\n",
      "||             other in the input (due to read sorting by location or \u001b[0m        ||\n",
      "||             reporting of multi-mapping read pairs). \u001b[0m                       ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    Pairing up the read pairs. \u001b[0m                                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||    Total alignments : \u001b[36m37474676\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m3259125 (8.7%)\u001b[0m \u001b[0m                      ||\n",
      "||    Running time : \u001b[36m2.89 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"count_pca/count_pca_old \u001b[0m ||\n",
      "|| .txt.summary\" \u001b[0m                                                             ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!featureCounts -p -B -C -M -O -T 16 -F SAF -a 'count_pca/count_pca_old.SAF' \\\n",
    "-o 'count_pca/count_pca_old.txt' \\\n",
    "BSF_01/res4/BSF_01/BSF_01_sorted.bam \\\n",
    "BSF_03/res4/BSF_03/BSF_03_sorted.bam \\\n",
    "BSF_04/res4/BSF_04/BSF_04_sorted.bam \\\n",
    "BSF_05/res4/BSF_05/BSF_05_sorted.bam \\\n",
    "BSF_06/res4/BSF_06/BSF_06_sorted.bam \\\n",
    "BSF_07/res4/BSF_07/BSF_07_sorted.bam \\\n",
    "BSF_08/res4/BSF_08/BSF_08_sorted.bam "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a5714d29-d6a6-4b53-a200-e4e5e0fcb053",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>BSF_01/res4/BSF_01/BSF_01_sorted.bam</th>\n",
       "      <th>BSF_03/res4/BSF_03/BSF_03_sorted.bam</th>\n",
       "      <th>BSF_04/res4/BSF_04/BSF_04_sorted.bam</th>\n",
       "      <th>BSF_05/res4/BSF_05/BSF_05_sorted.bam</th>\n",
       "      <th>BSF_06/res4/BSF_06/BSF_06_sorted.bam</th>\n",
       "      <th>BSF_07/res4/BSF_07/BSF_07_sorted.bam</th>\n",
       "      <th>BSF_08/res4/BSF_08/BSF_08_sorted.bam</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Tb927.8.4730:mRNA_1</th>\n",
       "      <td>Tb927_08_v5.1</td>\n",
       "      <td>1387684</td>\n",
       "      <td>1389085</td>\n",
       "      <td>.</td>\n",
       "      <td>1402</td>\n",
       "      <td>3320</td>\n",
       "      <td>630</td>\n",
       "      <td>134</td>\n",
       "      <td>140</td>\n",
       "      <td>4695</td>\n",
       "      <td>4461</td>\n",
       "      <td>4976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.7.5460:mRNA_2</th>\n",
       "      <td>Tb927_07_v5.1</td>\n",
       "      <td>1475328</td>\n",
       "      <td>1475778</td>\n",
       "      <td>.</td>\n",
       "      <td>451</td>\n",
       "      <td>58</td>\n",
       "      <td>73</td>\n",
       "      <td>54</td>\n",
       "      <td>110</td>\n",
       "      <td>32</td>\n",
       "      <td>50</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.4.580:mRNA_1</th>\n",
       "      <td>Tb927_04_v5.1</td>\n",
       "      <td>166931</td>\n",
       "      <td>167193</td>\n",
       "      <td>.</td>\n",
       "      <td>263</td>\n",
       "      <td>1235</td>\n",
       "      <td>372</td>\n",
       "      <td>428</td>\n",
       "      <td>49</td>\n",
       "      <td>1354</td>\n",
       "      <td>607</td>\n",
       "      <td>1073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.5.2260:mRNA_2</th>\n",
       "      <td>Tb927_05_v5.1</td>\n",
       "      <td>762326</td>\n",
       "      <td>762699</td>\n",
       "      <td>.</td>\n",
       "      <td>374</td>\n",
       "      <td>402</td>\n",
       "      <td>829</td>\n",
       "      <td>383</td>\n",
       "      <td>196</td>\n",
       "      <td>1375</td>\n",
       "      <td>952</td>\n",
       "      <td>1471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.4.740:mRNA_1</th>\n",
       "      <td>Tb927_04_v5.1</td>\n",
       "      <td>205296</td>\n",
       "      <td>205462</td>\n",
       "      <td>.</td>\n",
       "      <td>167</td>\n",
       "      <td>206</td>\n",
       "      <td>70</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1180</td>\n",
       "      <td>767</td>\n",
       "      <td>971</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               Chr    Start      End Strand  Length  \\\n",
       "Geneid                                                                \n",
       "Tb927.8.4730:mRNA_1  Tb927_08_v5.1  1387684  1389085      .    1402   \n",
       "Tb927.7.5460:mRNA_2  Tb927_07_v5.1  1475328  1475778      .     451   \n",
       "Tb927.4.580:mRNA_1   Tb927_04_v5.1   166931   167193      .     263   \n",
       "Tb927.5.2260:mRNA_2  Tb927_05_v5.1   762326   762699      .     374   \n",
       "Tb927.4.740:mRNA_1   Tb927_04_v5.1   205296   205462      .     167   \n",
       "\n",
       "                     BSF_01/res4/BSF_01/BSF_01_sorted.bam  \\\n",
       "Geneid                                                      \n",
       "Tb927.8.4730:mRNA_1                                  3320   \n",
       "Tb927.7.5460:mRNA_2                                    58   \n",
       "Tb927.4.580:mRNA_1                                   1235   \n",
       "Tb927.5.2260:mRNA_2                                   402   \n",
       "Tb927.4.740:mRNA_1                                    206   \n",
       "\n",
       "                     BSF_03/res4/BSF_03/BSF_03_sorted.bam  \\\n",
       "Geneid                                                      \n",
       "Tb927.8.4730:mRNA_1                                   630   \n",
       "Tb927.7.5460:mRNA_2                                    73   \n",
       "Tb927.4.580:mRNA_1                                    372   \n",
       "Tb927.5.2260:mRNA_2                                   829   \n",
       "Tb927.4.740:mRNA_1                                     70   \n",
       "\n",
       "                     BSF_04/res4/BSF_04/BSF_04_sorted.bam  \\\n",
       "Geneid                                                      \n",
       "Tb927.8.4730:mRNA_1                                   134   \n",
       "Tb927.7.5460:mRNA_2                                    54   \n",
       "Tb927.4.580:mRNA_1                                    428   \n",
       "Tb927.5.2260:mRNA_2                                   383   \n",
       "Tb927.4.740:mRNA_1                                      2   \n",
       "\n",
       "                     BSF_05/res4/BSF_05/BSF_05_sorted.bam  \\\n",
       "Geneid                                                      \n",
       "Tb927.8.4730:mRNA_1                                   140   \n",
       "Tb927.7.5460:mRNA_2                                   110   \n",
       "Tb927.4.580:mRNA_1                                     49   \n",
       "Tb927.5.2260:mRNA_2                                   196   \n",
       "Tb927.4.740:mRNA_1                                      2   \n",
       "\n",
       "                     BSF_06/res4/BSF_06/BSF_06_sorted.bam  \\\n",
       "Geneid                                                      \n",
       "Tb927.8.4730:mRNA_1                                  4695   \n",
       "Tb927.7.5460:mRNA_2                                    32   \n",
       "Tb927.4.580:mRNA_1                                   1354   \n",
       "Tb927.5.2260:mRNA_2                                  1375   \n",
       "Tb927.4.740:mRNA_1                                   1180   \n",
       "\n",
       "                     BSF_07/res4/BSF_07/BSF_07_sorted.bam  \\\n",
       "Geneid                                                      \n",
       "Tb927.8.4730:mRNA_1                                  4461   \n",
       "Tb927.7.5460:mRNA_2                                    50   \n",
       "Tb927.4.580:mRNA_1                                    607   \n",
       "Tb927.5.2260:mRNA_2                                   952   \n",
       "Tb927.4.740:mRNA_1                                    767   \n",
       "\n",
       "                     BSF_08/res4/BSF_08/BSF_08_sorted.bam  \n",
       "Geneid                                                     \n",
       "Tb927.8.4730:mRNA_1                                  4976  \n",
       "Tb927.7.5460:mRNA_2                                    45  \n",
       "Tb927.4.580:mRNA_1                                   1073  \n",
       "Tb927.5.2260:mRNA_2                                  1471  \n",
       "Tb927.4.740:mRNA_1                                    971  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "indf =pd.read_csv('count_pca/count_pca_old.txt',index_col=[0],comment='#',sep='\\t')\n",
    "indf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b78dea30-892c-4af1-8c5e-a7cb78c61701",
   "metadata": {},
   "outputs": [],
   "source": [
    "indf=indf.iloc[:,5:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c530c5b1-e903-43dd-8818-5a1e7df33bf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "indf.columns = ['Control','BLA_1','BLA_2','BLA_3','GAN_1','GAN_2','GAN_3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d06e03eb-be8a-45d5-b16b-a7829cdaec66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Control</th>\n",
       "      <th>BLA_1</th>\n",
       "      <th>BLA_2</th>\n",
       "      <th>BLA_3</th>\n",
       "      <th>GAN_1</th>\n",
       "      <th>GAN_2</th>\n",
       "      <th>GAN_3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Tb927.8.4730:mRNA_1</th>\n",
       "      <td>3320</td>\n",
       "      <td>630</td>\n",
       "      <td>134</td>\n",
       "      <td>140</td>\n",
       "      <td>4695</td>\n",
       "      <td>4461</td>\n",
       "      <td>4976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.7.5460:mRNA_2</th>\n",
       "      <td>58</td>\n",
       "      <td>73</td>\n",
       "      <td>54</td>\n",
       "      <td>110</td>\n",
       "      <td>32</td>\n",
       "      <td>50</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.4.580:mRNA_1</th>\n",
       "      <td>1235</td>\n",
       "      <td>372</td>\n",
       "      <td>428</td>\n",
       "      <td>49</td>\n",
       "      <td>1354</td>\n",
       "      <td>607</td>\n",
       "      <td>1073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.5.2260:mRNA_2</th>\n",
       "      <td>402</td>\n",
       "      <td>829</td>\n",
       "      <td>383</td>\n",
       "      <td>196</td>\n",
       "      <td>1375</td>\n",
       "      <td>952</td>\n",
       "      <td>1471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tb927.4.740:mRNA_1</th>\n",
       "      <td>206</td>\n",
       "      <td>70</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1180</td>\n",
       "      <td>767</td>\n",
       "      <td>971</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Control  BLA_1  BLA_2  BLA_3  GAN_1  GAN_2  GAN_3\n",
       "Geneid                                                                \n",
       "Tb927.8.4730:mRNA_1     3320    630    134    140   4695   4461   4976\n",
       "Tb927.7.5460:mRNA_2       58     73     54    110     32     50     45\n",
       "Tb927.4.580:mRNA_1      1235    372    428     49   1354    607   1073\n",
       "Tb927.5.2260:mRNA_2      402    829    383    196   1375    952   1471\n",
       "Tb927.4.740:mRNA_1       206     70      2      2   1180    767    971"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "502de633-4ea0-49f3-b0d4-bef7680e1003",
   "metadata": {},
   "outputs": [],
   "source": [
    "colors = ['k','g','g','g','m','m','m']\n",
    "color_to_label = {'k':'plasmid library','g':'BSD','r':'GCV'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "99504a3a-04f6-4170-b484-9f0a6c63fc9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.manifold import MDS\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from adjustText import adjust_text\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "def normalize_dataframe(in_df):\n",
    "    sc = StandardScaler()\n",
    "    return pd.DataFrame(sc.fit_transform(in_df),\n",
    "                        index=in_df.index, \n",
    "                        columns=in_df.columns)\n",
    "\n",
    "def plot_pca_columns(in_df, colors, color_to_label, figsize=(4,4),annotate=True,adjust=True, fig_name = False):\n",
    "    # Normalize the input DataFrame\n",
    "    normalized_df = normalize_dataframe(in_df)\n",
    "\n",
    "    # Perform PCA on the transposed DataFrame (to analyze columns)\n",
    "    pca = PCA(n_components=2)\n",
    "    pca_transformed_columns = pca.fit_transform(normalized_df.T)\n",
    "\n",
    "    # Create a scatter plot of the PCA-transformed column coordinates\n",
    "    fig, ax = plt.subplots(figsize=figsize)\n",
    "    for i, color in enumerate(colors):\n",
    "        ax.scatter(pca_transformed_columns[i, 0], pca_transformed_columns[i, 1],\n",
    "                   alpha=0.5, c=color, marker='o')\n",
    "\n",
    "    if annotate:\n",
    "        # Add column labels\n",
    "        texts = []\n",
    "        for i, column_name in enumerate(in_df.columns):\n",
    "            texts.append(ax.text(pca_transformed_columns[i, 0], pca_transformed_columns[i, 1], \n",
    "                                 column_name, fontsize=12, color='darkslategray'))\n",
    "\n",
    "        if adjust:\n",
    "            # Adjust the text labels to avoid overlaps\n",
    "            adjust_text(texts, arrowprops=dict(arrowstyle='->', color='red'))\n",
    "\n",
    "    # Create legend\n",
    "    legend_elements = [plt.Line2D([0], [0], linestyle='', marker='o', color=color, label=label, markersize=8)\n",
    "                       for color, label in color_to_label.items()]\n",
    "    ax.legend(handles=legend_elements, loc='center left', bbox_to_anchor=(1, 0.5))\n",
    "\n",
    "    # Set plot labels and title\n",
    "    ax.set_xlabel(f'PC1 ({pca.explained_variance_ratio_[0] * 100:.2f}%)', fontsize=10)\n",
    "    ax.set_ylabel(f'PC2 ({pca.explained_variance_ratio_[1] * 100:.2f}%)', fontsize=10)\n",
    "    ax.set_title('PCA', fontsize=12)\n",
    "\n",
    "    # Customize axis spines\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['left'].set_position(('outward', 10))\n",
    "    ax.spines['bottom'].set_position(('outward', 10))\n",
    "    \n",
    "    if fig_name:\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(fig_name)\n",
    "        \n",
    "    \n",
    "    pca_transformed_columns = pd.DataFrame(pca_transformed_columns)\n",
    "    pca_transformed_columns.columns = ['DIM_1','DIM_2']\n",
    "    pca_transformed_columns.index = normalized_df.columns\n",
    "    # Show the plot\n",
    "    # plt.show()\n",
    "    return pca_transformed_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "995f874b-2263-4f73-95fc-b1832580a4a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "pca_values = plot_pca_columns(indf, colors, color_to_label, figsize=(6,5), annotate=False, adjust=False, fig_name = 'count_pca/pca_old.png')\n",
    "pca_values.to_csv('count_pca/pca_old.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69cc29cb-f796-4fec-9ea4-9d23ed2334db",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:miniconda3-ritSeq]",
   "language": "python",
   "name": "conda-env-miniconda3-ritSeq-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
